From 7c7cabdd04c048ab106db8975f0a7523a93b88e4 Mon Sep 17 00:00:00 2001
From: Gopi Kumar Bulusu <gopi@sankhya.com>
Date: Tue, 11 Aug 2026 18:50:09 +0530
Subject: [PATCH 2/2] Add MicroBlaze 64-bit support

Added support for new MicroBlaze 64-bit architecture under gas/
ld/ and binutils/. New 64-bit tests added for gas and ld.

Signed-off-by :Nagaraju Mekala <nmekala@xilix.com>
Signed-off-by :Mahesh Bodapati <mbodapat@xilinx.com>
Signed-off-by: Gopi Kumar Bulusu <gopi@sankhya.com>

Upstream-Status: Pending
---
 gas/config/tc-microblaze.c                  | 1084 +++++++++++++++++--
 gas/config/tc-microblaze.h                  |    4 +-
 gas/testsuite/gas/microblaze/allinsn64.d    |  252 +++++
 gas/testsuite/gas/microblaze/allinsn64.exp  |   21 +
 gas/testsuite/gas/microblaze/allinsn64.s    |  234 ++++
 gas/testsuite/gas/microblaze/brea.d         |  132 +++
 gas/testsuite/gas/microblaze/brea.exp       |   21 +
 gas/testsuite/gas/microblaze/brea.s         |  129 +++
 gas/testsuite/gas/microblaze/reloc_64.d     |   23 +
 gas/testsuite/gas/microblaze/reloc_64.exp   |   21 +
 gas/testsuite/gas/microblaze/reloc_64.s     |   26 +
 ld/Makefile.am                              |    2 +
 ld/configure.tgt                            |    5 +-
 ld/emulparams/elf64microblaze.sh            |   23 +
 ld/emulparams/elf64microblazeel.sh          |   23 +
 ld/testsuite/ld-microblaze/microblaze64.exp |   26 +
 ld/testsuite/ld-microblaze/s64.d            |   66 ++
 ld/testsuite/ld-microblaze/s64.s            |   82 ++
 ld/testsuite/ld-microblaze/s64defs.s        |   27 +
 ld/testsuite/ld-microblaze/simple.ld        |    9 +
 20 files changed, 2113 insertions(+), 97 deletions(-)
 create mode 100644 gas/testsuite/gas/microblaze/allinsn64.d
 create mode 100644 gas/testsuite/gas/microblaze/allinsn64.exp
 create mode 100644 gas/testsuite/gas/microblaze/allinsn64.s
 create mode 100644 gas/testsuite/gas/microblaze/brea.d
 create mode 100644 gas/testsuite/gas/microblaze/brea.exp
 create mode 100644 gas/testsuite/gas/microblaze/brea.s
 create mode 100644 gas/testsuite/gas/microblaze/reloc_64.d
 create mode 100644 gas/testsuite/gas/microblaze/reloc_64.exp
 create mode 100644 gas/testsuite/gas/microblaze/reloc_64.s
 create mode 100644 ld/emulparams/elf64microblaze.sh
 create mode 100644 ld/emulparams/elf64microblazeel.sh
 create mode 100644 ld/testsuite/ld-microblaze/microblaze64.exp
 create mode 100644 ld/testsuite/ld-microblaze/s64.d
 create mode 100644 ld/testsuite/ld-microblaze/s64.s
 create mode 100644 ld/testsuite/ld-microblaze/s64defs.s
 create mode 100644 ld/testsuite/ld-microblaze/simple.ld

diff --git a/gas/config/tc-microblaze.c b/gas/config/tc-microblaze.c
index a7d410f879b..df1a8bf5c45 100644
--- a/gas/config/tc-microblaze.c
+++ b/gas/config/tc-microblaze.c
@@ -35,10 +35,13 @@
 #define streq(a,b) (strcmp (a, b) == 0)
 #endif
 
+static int microblaze_arch_size = 0;
+
 #define OPTION_EB (OPTION_MD_BASE + 0)
 #define OPTION_EL (OPTION_MD_BASE + 1)
 #define OPTION_LITTLE (OPTION_MD_BASE + 2)
 #define OPTION_BIG (OPTION_MD_BASE + 3)
+#define OPTION_M64 (OPTION_MD_BASE + 4)
 
 void microblaze_generate_symbol (char *sym);
 static bool check_spl_reg (unsigned *);
@@ -91,6 +94,8 @@ const char FLT_CHARS[] = "rRsSfFdDxXpP";
 #define TLSTPREL_OFFSET      16
 #define TEXT_OFFSET	     17
 #define TEXT_PC_OFFSET       18
+#define DEFINED_64_OFFSET   19
+#define DEFINED_64_PC_OFFSET 20
 
 /* Initialize the relax table.  */
 const relax_typeS md_relax_table[] =
@@ -113,7 +118,9 @@ const relax_typeS md_relax_table[] =
   { 0x7fffffff, 0x80000000, INST_WORD_SIZE*2, 0 },  /* 15: TLSGOTTPREL_OFFSET.  */
   { 0x7fffffff, 0x80000000, INST_WORD_SIZE*2, 0 },  /* 16: TLSTPREL_OFFSET.  */
   { 0x7fffffff, 0x80000000, INST_WORD_SIZE*2, 0 },  /* 17: TEXT_OFFSET.  */
-  { 0x7fffffff, 0x80000000, INST_WORD_SIZE*2, 0 }   /* 18: TEXT_PC_OFFSET.  */
+  { 0x7fffffff, 0x80000000, INST_WORD_SIZE*2, 0 },  /* 18: TEXT_PC_OFFSET.  */
+  { 0x7fffffffffffffff, 0x8000000000000000, INST_WORD_SIZE, 0 },   /* 19: DEFINED_64_OFFSET.  */
+  { 0x7fffffffffffffff, 0x8000000000000000, INST_WORD_SIZE*2, 0 }   /* 20: DEFINED_64_PC_OFFSET.  */
 };
 
 static htab_t  opcode_hash_control;	/* Opcode mnemonics.  */
@@ -351,6 +358,17 @@ microblaze_s_weakext (int ignore ATTRIBUTE_UNUSED)
   demand_empty_rest_of_line ();
 }
 
+/* Handle the .gpword pseudo-op, Pass to s_rva */
+
+static void
+microblaze_s_gpword (int ignore ATTRIBUTE_UNUSED)
+{
+  int size = 4;
+  if (microblaze_arch_size == 64)
+    size = 8;
+  s_rva(size);
+}
+
 /* This table describes all the machine specific pseudo-ops the assembler
    has to support.  The fields are:
    Pseudo-op name without dot
@@ -366,7 +384,7 @@ const pseudo_typeS md_pseudo_table[] =
   {"data32", cons, 4},     /* Same as word.  */
   {"ent", s_func, 0}, /* Treat ent as function entry point.  */
   {"end", microblaze_s_func, 1}, /* Treat end as function end point.  */
-  {"gpword", s_rva, 4}, /* gpword label => store resolved label address in data section.  */
+  {"gpword", microblaze_s_gpword, 0}, /* gpword label => store resolved label address in data section.  */
   {"weakext", microblaze_s_weakext, 0},
   {"rodata", microblaze_s_rdata, 0},
   {"sdata2", microblaze_s_rdata, 1},
@@ -376,6 +394,7 @@ const pseudo_typeS md_pseudo_table[] =
 #endif
   {"sbss", microblaze_s_sbss, 0},
   {"word", cons, 4},
+  {"dword", cons, 8},
   {"frame", s_ignore, 0},
   {"mask", s_ignore, 0}, /* Emitted by gcc.  */
   {NULL, NULL, 0}
@@ -387,13 +406,34 @@ const pseudo_typeS md_pseudo_table[] =
 void
 md_begin (void)
 {
-  const struct op_code_struct * opcode;
+  struct op_code_struct * opcode;
+  const char *prev_name = "";
 
   opcode_hash_control = str_htab_create ();
 
   /* Insert unique names into hash table.  */
   for (opcode = microblaze_opcodes; opcode->name; opcode ++)
-    str_hash_insert (opcode_hash_control, opcode->name, opcode, 0);
+    {
+      if (strcmp (prev_name, opcode->name))
+	{
+	  prev_name = (char *) opcode->name;
+	  str_hash_insert (opcode_hash_control, opcode->name, opcode, 0);
+	}
+    }
+}
+
+static int
+is_reg (char * s)
+{
+  int is_reg = 0;
+  /* Strip leading whitespace.  */
+  while (ISSPACE (* s))
+    ++ s;
+  if (TOLOWER (s[0]) == 'r')
+    {
+	 is_reg =1;
+    }
+  return is_reg;
 }
 
 /* Try to parse a reg name.  */
@@ -747,6 +787,74 @@ parse_imm (char * s, expressionS * e, offsetT min, offsetT max)
   return new_pointer;
 }
 
+ static char *
+parse_imml (char * s, expressionS * e, long long min, long long max)
+{
+  char *new_pointer;
+  char *atp;
+  int itype, ilen;
+
+  ilen = 0;
+
+  /* Find the start of "@GOT" or "@PLT" suffix (if any) */
+  for (atp = s; *atp != '@'; atp++)
+    if (is_end_of_stmt (*atp))
+      break;
+
+  if (*atp == '@')
+    {
+      itype = match_imm (atp + 1, &ilen);
+      if (itype != 0)
+	{
+	  *atp = 0;
+	  e->X_md = itype;
+	}
+      else
+	{
+	  atp = NULL;
+	  e->X_md = 0;
+	  ilen = 0;
+	}
+      *atp = 0;
+    }
+  else
+    {
+      atp = NULL;
+      e->X_md = 0;
+    }
+
+  if (atp && !GOT_symbol)
+    {
+      GOT_symbol = symbol_find_or_make (GOT_SYMBOL_NAME);
+    }
+
+  new_pointer = parse_exp (s, e);
+
+  if (!GOT_symbol && ! strncmp (s, GOT_SYMBOL_NAME, 20))
+    {
+      GOT_symbol = symbol_find_or_make (GOT_SYMBOL_NAME);
+    }
+
+  if (e->X_op == O_absent)
+    ; /* An error message has already been emitted.  */
+  else if ((e->X_op != O_constant && e->X_op != O_symbol) )
+    as_fatal (_("operand must be a constant or a label"));
+  else if ((e->X_op == O_constant) && ((long long) e->X_add_number < min
+				       || (long long) e->X_add_number > max))
+    {
+      as_fatal (_("operand must be absolute in range %lld..%lld, not %lld"),
+		min, max, (long long) e->X_add_number);
+    }
+
+  if (atp)
+    {
+      *atp = '@'; /* restore back (needed?)  */
+      if (new_pointer >= atp)
+	new_pointer += ilen + 1; /* sizeof (imm_suffix) + 1 for '@' */
+    }
+  return new_pointer;
+}
+
 static char *
 check_got (int * got_type, int * got_len)
 {
@@ -801,7 +909,7 @@ check_got (int * got_type, int * got_len)
 extern bfd_reloc_code_real_type
 parse_cons_expression_microblaze (expressionS *exp, int size)
 {
-  if (size == 4)
+  if (size == 4 || (microblaze_arch_size == 64 && size == 8))
     {
       /* Handle @GOTOFF et.al.  */
       char *save, *gotfree_copy;
@@ -833,6 +941,7 @@ parse_cons_expression_microblaze (expressionS *exp, int size)
 
 static const char * str_microblaze_ro_anchor = "RO";
 static const char * str_microblaze_rw_anchor = "RW";
+static const char * str_microblaze_64 = "64";
 
 static bool
 check_spl_reg (unsigned * reg)
@@ -882,7 +991,8 @@ md_assemble (char * str)
 {
   char * op_start;
   char * op_end;
-  struct op_code_struct * opcode, *opcode1;
+  char * temp_op_end;
+  struct op_code_struct * opcode, *opcode1, *opcode2;
   char * output = NULL;
   int nlen = 0;
   int i;
@@ -890,10 +1000,13 @@ md_assemble (char * str)
   unsigned reg1;
   unsigned reg2;
   unsigned reg3;
+  unsigned rfslreg;
   unsigned isize;
-  unsigned int immed = 0, immed2 = 0, temp;
+  unsigned long long immed = 0, immed2 = 0, temp;
   expressionS exp;
   char name[20];
+  long immedl;
+  int reg=0;
 
   /* Drop leading whitespace.  */
   while (is_whitespace (* str))
@@ -925,6 +1038,77 @@ md_assemble (char * str)
       return;
     }
 
+  if ((microblaze_arch_size == 64) && (streq (name, "addli") || streq (name, "addlic") ||
+	streq (name, "addlik") || streq (name, "addlikc") || streq (name, "rsubli")
+	|| streq (name, "rsublic") || streq (name, "rsublik") || streq (name, "rsublikc")
+	|| streq (name, "andli") || streq (name, "andnli") || streq (name, "orli")
+	|| streq (name, "xorli")))
+    {
+      temp_op_end = op_end;
+      if (strcmp (temp_op_end, ""))
+	temp_op_end = parse_reg (temp_op_end + 1, &reg1);  /* Get rd.  */
+      if (strcmp (temp_op_end, ""))
+	reg  = is_reg (temp_op_end + 1);
+      if (reg)
+	{
+
+	  opcode->inst_type=INST_TYPE_RD_R1_IMML;
+	  opcode->inst_offset_type = INST_NO_OFFSET;
+	  if (streq (name, "addli"))
+	    opcode->bit_sequence = ADDLI_MASK;
+	  else if (streq (name, "addlic"))
+	    opcode->bit_sequence = ADDLIC_MASK;
+	  else if (streq (name, "addlik"))
+	    opcode->bit_sequence = ADDLIK_MASK;
+	  else if (streq (name, "addlikc"))
+	    opcode->bit_sequence = ADDLIKC_MASK;
+	  else if (streq (name, "rsubli"))
+	    opcode->bit_sequence = RSUBLI_MASK;
+	  else if (streq (name, "rsublic"))
+	    opcode->bit_sequence = RSUBLIC_MASK;
+	  else if (streq (name, "rsublik"))
+	    opcode->bit_sequence = RSUBLIK_MASK;
+	  else if (streq (name, "rsublikc"))
+	    opcode->bit_sequence = RSUBLIKC_MASK;
+	  else if (streq (name, "andli"))
+	    opcode->bit_sequence = ANDLI_MASK;
+	  else if (streq (name, "andnli"))
+	    opcode->bit_sequence = ANDLNI_MASK;
+	  else if (streq (name, "orli"))
+	    opcode->bit_sequence = ORLI_MASK;
+	  else if (streq (name, "xorli"))
+	    opcode->bit_sequence = XORLI_MASK;
+	}
+      else
+	{
+	  opcode->inst_type=INST_TYPE_RD_IMML;
+	  opcode->inst_offset_type = INST_NO_OFFSET;
+	  if (streq (name, "addli"))
+	    opcode->bit_sequence = ADDLI_ONE_REG_MASK;
+	  else if (streq (name, "addlic"))
+	    opcode->bit_sequence = ADDLIC_ONE_REG_MASK;
+	  else if (streq (name, "addlik"))
+	    opcode->bit_sequence = ADDLIK_ONE_REG_MASK;
+	  else if (streq (name, "addlikc"))
+	    opcode->bit_sequence = ADDLIKC_ONE_REG_MASK;
+	  else if (streq (name, "rsubli"))
+	    opcode->bit_sequence = RSUBLI_ONE_REG_MASK;
+	  else if (streq (name, "rsublic"))
+	    opcode->bit_sequence = RSUBLIC_ONE_REG_MASK;
+	  else if (streq (name, "rsublik"))
+	    opcode->bit_sequence = RSUBLIK_ONE_REG_MASK;
+	  else if (streq (name, "rsublikc"))
+	    opcode->bit_sequence = RSUBLIKC_ONE_REG_MASK;
+	  else if (streq (name, "andli"))
+	    opcode->bit_sequence = ANDLI_ONE_REG_MASK;
+	  else if (streq (name, "andnli"))
+	    opcode->bit_sequence = ANDLNI_ONE_REG_MASK;
+	  else if (streq (name, "orli"))
+	    opcode->bit_sequence = ORLI_ONE_REG_MASK;
+	  else if (streq (name, "xorli"))
+	    opcode->bit_sequence = XORLI_ONE_REG_MASK;
+	}
+    }
   inst = opcode->bit_sequence;
   isize = 4;
 
@@ -993,7 +1177,12 @@ md_assemble (char * str)
           reg2 = 0;
         }
       if (strcmp (op_end, ""))
+	{
+	if(microblaze_arch_size == 64)
+	op_end = parse_imml (op_end + 1, & exp, MIN_IMML, MAX_IMML);
+	else
 	op_end = parse_imm (op_end + 1, & exp, MIN_IMM, MAX_IMM);
+	}
       else
 	as_fatal (_("Error in statement syntax"));
 
@@ -1012,8 +1201,16 @@ md_assemble (char * str)
 	    as_fatal (_("lmi pseudo instruction should not use a label in imm field"));
 	  else if (streq (name, "smi"))
 	    as_fatal (_("smi pseudo instruction should not use a label in imm field"));
-
-	  if (reg2 == REG_ROSDP)
+	  if(streq (name, "lli") || streq (name, "sli"))
+	    opc = str_microblaze_64;
+	  else if ((microblaze_arch_size == 64) && ((streq (name, "lbui")
+			|| streq (name, "lhui") || streq (name, "lwi") || streq (name, "sbi")
+			|| streq (name, "shi") || streq (name, "swi"))))
+	    {
+	      opc = str_microblaze_64;
+	      subtype = opcode->inst_offset_type;
+	    }
+	  else if (reg2 == REG_ROSDP)
 	    opc = str_microblaze_ro_anchor;
 	  else if (reg2 == REG_RWSDP)
 	    opc = str_microblaze_rw_anchor;
@@ -1076,34 +1273,88 @@ md_assemble (char * str)
               inst |= (immed << IMM_LOW) & IMM_MASK;
             }
 	}
-      else
+      else if (streq (name, "lli") || streq (name, "sli") || ((microblaze_arch_size == 64)
+		&& ((streq (name, "lbui")) || streq (name, "lhui")
+		|| streq (name, "lwi") || streq (name, "sbi")
+		|| streq (name, "shi") || streq (name, "swi"))))
 	{
-          temp = immed & 0xFFFF8000;
-          if ((temp != 0) && (temp != 0xFFFF8000))
+	  temp = ((long long)immed) & 0xFFFFFFFFFFFF8000;
+	  if (temp != 0 && temp != 0xFFFFFFFFFFFF8000)
 	    {
-              /* Needs an immediate inst.  */
-	      opcode1 = str_hash_find (opcode_hash_control, "imm");
-              if (opcode1 == NULL)
+	      /* Needs an immediate inst.  */
+	      if (((long long)immed) > (long long)-549755813888 && ((long long)immed) < (long long)549755813887)
+		{
+		  opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+		  if (opcode1 == NULL)
+		    {
+		      as_bad (_("unknown opcode \"%s\""), "imml");
+		      return;
+		    }
+		  inst1 = opcode1->bit_sequence;
+		  inst1 |= ((immed & 0xFFFFFFFFFFFF0000L) >> 16) & IMML_MASK;
+		  output[0] = INST_BYTE0 (inst1);
+		  output[1] = INST_BYTE1 (inst1);
+		  output[2] = INST_BYTE2 (inst1);
+		  output[3] = INST_BYTE3 (inst1);
+		  output = frag_more (isize);
+	    }
+	  else
+	   {
+	   opcode2 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	   opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      if (opcode1 == NULL || opcode2 == NULL)
                 {
-                  as_bad (_("unknown opcode \"%s\""), "imm");
+		  as_bad (_("unknown opcode \"%s\""), "imml");
                   return;
                 }
-
+	      inst1 = opcode2->bit_sequence;
+	      inst1 |= ((immed & 0xFFFFFF0000000000L) >> 40) & IMML_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
               inst1 = opcode1->bit_sequence;
-              inst1 |= ((immed & 0xFFFF0000) >> 16) & IMM_MASK;
+	      inst1 |= ((immed & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
               output[0] = INST_BYTE0 (inst1);
               output[1] = INST_BYTE1 (inst1);
               output[2] = INST_BYTE2 (inst1);
               output[3] = INST_BYTE3 (inst1);
               output = frag_more (isize);
 	    }
+	  }
+	  inst |= (reg1 << RD_LOW) & RD_MASK;
+	  inst |= (reg2 << RA_LOW) & RA_MASK;
+	  inst |= (immed << IMM_LOW) & IMM_MASK;
+	}
+      else
+	{
+	  temp = immed & 0xFFFF8000;
+	  if ((temp != 0) && (temp != 0xFFFF8000))
+	    {
+	      /* Needs an immediate inst.  */
+	      opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imm");
+	      if (opcode1 == NULL)
+		{
+		  as_bad (_("unknown opcode \"%s\""), "imm");
+		  return;
+		}
+
+	      inst1 = opcode1->bit_sequence;
+	      inst1 |= ((immed & 0xFFFF0000) >> 16) & IMM_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
+	    }
 	  inst |= (reg1 << RD_LOW) & RD_MASK;
 	  inst |= (reg2 << RA_LOW) & RA_MASK;
 	  inst |= (immed << IMM_LOW) & IMM_MASK;
 	}
       break;
 
-    case INST_TYPE_RD_R1_IMM5:
+    case INST_TYPE_RD_R1_IMMS:
       if (strcmp (op_end, ""))
         op_end = parse_reg (op_end + 1, &reg1);  /* Get rd.  */
       else
@@ -1137,14 +1388,20 @@ md_assemble (char * str)
           immed = exp.X_add_number;
         }
 
-      if (immed != (immed % 32))
+      if ((immed != (immed % 32)) &&
+	(opcode->instr == bslli || opcode->instr == bsrai || opcode->instr == bsrli))
 	{
           as_warn (_("Shift value > 32. using <value %% 32>"));
           immed = immed % 32;
         }
+      else if (immed != (immed % 64))
+	{
+	  as_warn (_("Shift value > 64. using <value %% 64>"));
+	  immed = immed % 64;
+	}
       inst |= (reg1 << RD_LOW) & RD_MASK;
       inst |= (reg2 << RA_LOW) & RA_MASK;
-      inst |= (immed << IMM_LOW) & IMM5_MASK;
+      inst |= (immed << IMM_LOW) & IMM6_MASK;
       break;
 
     case INST_TYPE_RD_R1_IMMW_IMMS:
@@ -1172,7 +1429,7 @@ md_assemble (char * str)
 
       /* Width immediate value.  */
       if (strcmp (op_end, ""))
-	op_end = parse_imm (op_end + 1, &exp, MIN_IMM_WIDTH, MAX_IMM_WIDTH);
+	op_end = parse_imm (op_end + 1, &exp, MIN_IMM, MAX_IMM);
       else
 	as_fatal (_("Error in statement syntax"));
 
@@ -1187,6 +1444,8 @@ md_assemble (char * str)
 
       if (opcode->instr == bsefi && immed > 31)
 	as_fatal (_("Width value must be less than 32"));
+      else if (opcode->instr == bslefi && immed > 63)
+	as_fatal (_("Width value must be less than 64"));
 
       /* Shift immediate value.  */
       if (strcmp (op_end, ""))
@@ -1206,26 +1465,32 @@ md_assemble (char * str)
 	  immed2 = exp.X_add_number;
 	}
 
-      if (immed2 != (immed2 % 32))
+      if ((immed2 != (immed2 % 32)) && (opcode->instr == bsefi || opcode->instr == bsifi))
 	{
 	  as_warn (_("Shift value greater than 32. using <value %% 32>"));
 	  immed2 = immed2 % 32;
 	}
+      else if (immed2 != (immed2 % 64))
+	{
+	  as_warn (_("Shift value greater than 64. using <value %% 64>"));
+	  immed2 = immed2 % 64;
+	}
 
       /* Check combined value.  */
-      if (immed + immed2 > 32)
+      if ((immed + immed2 > 32) && (opcode->instr == bsefi || opcode->instr == bsifi))
 	as_fatal (_("Width value + shift value must not be greater than 32"));
+      else if (immed + immed2 > 64)
+	as_fatal (_("Width value + shift value must not be greater than 64"));
 
       inst |= (reg1 << RD_LOW) & RD_MASK;
       inst |= (reg2 << RA_LOW) & RA_MASK;
 
-      if (opcode->instr == bsefi)
-	inst |= (immed & IMM5_MASK) << IMM_WIDTH_LOW; /* bsefi */
+      if (opcode->instr == bsefi || opcode->instr == bslefi)
+	inst |= (immed & IMM6_MASK) << IMM_WIDTH_LOW; /* bsefi or bslefi */
       else
-	inst |= ((immed + immed2 - 1) & IMM5_MASK)
-		<< IMM_WIDTH_LOW; /* bsifi */
+	inst |= ((immed + immed2 - 1) & IMM6_MASK) << IMM_WIDTH_LOW; /* bsifi or bslifi */
 
-      inst |= (immed2 << IMM_LOW) & IMM5_MASK;
+      inst |= (immed2 << IMM_LOW) & IMM6_MASK;
       break;
 
     case INST_TYPE_R1_R2:
@@ -1291,7 +1556,7 @@ md_assemble (char * str)
           reg1 = 0;
         }
       if (strcmp (op_end, ""))
-        op_end = parse_reg (op_end + 1, &immed);  /* Get rfslN.  */
+	op_end = parse_reg (op_end + 1, &rfslreg);  /* Get rfslN.  */
       else
 	{
           as_fatal (_("Error in statement syntax"));
@@ -1303,7 +1568,7 @@ md_assemble (char * str)
         as_fatal (_("Cannot use special register with this instruction"));
 
       inst |= (reg1 << RD_LOW) & RD_MASK;
-      inst |= (immed << IMM_LOW) & RFSL_MASK;
+      inst |= (rfslreg << IMM_LOW) & RFSL_MASK;
       output = frag_more (isize);
       break;
 
@@ -1326,7 +1591,7 @@ md_assemble (char * str)
         as_fatal (_("Cannot use special register with this instruction"));
 
       if (exp.X_op != O_constant)
-        as_fatal (_("Symbol used as immediate value for msrset/msrclr instructions"));
+	as_fatal (_("Symbol used as immediate value for arithmetic long instructions"));
       else
 	{
           output = frag_more (isize);
@@ -1336,6 +1601,95 @@ md_assemble (char * str)
       inst |= (immed << IMM_LOW) & IMM15_MASK;
       break;
 
+    case INST_TYPE_RD_IMML:
+      if (strcmp (op_end, ""))
+	op_end = parse_reg (op_end + 1, &reg1);  /* Get rd.  */
+      else
+	{
+	  as_fatal (_("Error in statement syntax"));
+	  reg1 = 0;
+	}
+
+      if (strcmp (op_end, ""))
+	op_end = parse_imml (op_end + 1, & exp, MIN_IMML, MAX_IMML);
+      else
+	as_fatal (_("Error in statement syntax"));
+
+      /* Check for spl registers. */
+      if (check_spl_reg (&reg1))
+	as_fatal (_("Cannot use special register with this instruction"));
+      if (exp.X_op != O_constant)
+	{
+	  char *opc = NULL;
+	  relax_substateT subtype;
+
+	  if (exp.X_md != 0)
+	    subtype = get_imm_otype(exp.X_md);
+	  else
+	    subtype = opcode->inst_offset_type;
+
+	  output = frag_var (rs_machine_dependent,
+			     isize * 2,
+			     isize * 2,
+			     subtype,
+			     exp.X_add_symbol,
+			     exp.X_add_number,
+			     (char *) opc);
+	  immed = 0L;
+	}
+      else
+	{
+	  output = frag_more (isize);
+	  immed = exp.X_add_number;
+	  temp = ((long long)immed) & 0xFFFFFFFFFFFF8000;
+	  if (temp != 0 && temp != 0xFFFFFFFFFFFF8000 && temp != 0x8000)
+	    {
+	      /* Needs an immediate inst.  */
+	   if (((long long)immed) > (long long)-549755813888 && ((long long)immed) < (long long)549755813887)
+	    {
+	    opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      if (opcode1 == NULL)
+		{
+		  as_bad (_("unknown opcode \"%s\""), "imml");
+		  return;
+		}
+	      inst1 = opcode1->bit_sequence;
+	      inst1 |= ((immed & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
+	    }
+	  else {
+	   opcode2 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	   opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      if (opcode1 == NULL || opcode2 == NULL)
+		{
+		  as_bad (_("unknown opcode \"%s\""), "imml");
+		  return;
+		}
+	      inst1 = opcode2->bit_sequence;
+	      inst1 |= ((immed & 0xFFFFFF0000000000L) >> 40) & IMML_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
+	      inst1 = opcode1->bit_sequence;
+	      inst1 |= ((immed & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
+	  }
+	  }
+	}
+      inst |= (reg1 << RD_LOW) & RD_MASK;
+      inst |= (immed << IMM_LOW) & IMM16_MASK;
+      break;
+
     case INST_TYPE_R1_RFSL:
       if (strcmp (op_end, ""))
         op_end = parse_reg (op_end + 1, &reg1);  /* Get r1.  */
@@ -1345,7 +1699,7 @@ md_assemble (char * str)
           reg1 = 0;
         }
       if (strcmp (op_end, ""))
-        op_end = parse_reg (op_end + 1, &immed);  /* Get rfslN.  */
+	op_end = parse_reg (op_end + 1, &rfslreg);  /* Get rfslN.  */
       else
 	{
           as_fatal (_("Error in statement syntax"));
@@ -1357,19 +1711,19 @@ md_assemble (char * str)
         as_fatal (_("Cannot use special register with this instruction"));
 
       inst |= (reg1 << RA_LOW) & RA_MASK;
-      inst |= (immed << IMM_LOW) & RFSL_MASK;
+      inst |= (rfslreg << IMM_LOW) & RFSL_MASK;
       output = frag_more (isize);
       break;
 
     case INST_TYPE_RFSL:
       if (strcmp (op_end, ""))
-        op_end = parse_reg (op_end + 1, &immed);  /* Get rfslN.  */
+	op_end = parse_reg (op_end + 1, &rfslreg);  /* Get rfslN.  */
       else
 	{
           as_fatal (_("Error in statement syntax"));
           immed = 0;
         }
-      inst |= (immed << IMM_LOW) & RFSL_MASK;
+      inst |= (rfslreg << IMM_LOW) & RFSL_MASK;
       output = frag_more (isize);
       break;
 
@@ -1645,7 +1999,12 @@ md_assemble (char * str)
           reg1 = 0;
         }
       if (strcmp (op_end, ""))
+      {
+	if(microblaze_arch_size == 64)
+	op_end = parse_imml (op_end + 1, & exp, MIN_IMML, MAX_IMML);
+	else
         op_end = parse_imm (op_end + 1, & exp, MIN_IMM, MAX_IMM);
+      }
       else
         as_fatal (_("Error in statement syntax"));
 
@@ -1655,11 +2014,16 @@ md_assemble (char * str)
 
       if (exp.X_op != O_constant)
 	{
-          char *opc = NULL;
+	  const char *opc = NULL;
           relax_substateT subtype;
 
 	  if (exp.X_md != 0)
 	    subtype = get_imm_otype(exp.X_md);
+	  else if (streq (name, "brealid") || streq (name, "breaid") || streq (name, "breai"))
+	    {
+	      opc = str_microblaze_64;
+	      subtype = opcode->inst_offset_type;
+	    }
 	  else
 	    subtype = opcode->inst_offset_type;
 
@@ -1669,7 +2033,7 @@ md_assemble (char * str)
 			     subtype,   /* PC-relative or not.  */
 			     exp.X_add_symbol,
 			     exp.X_add_number,
-			     opc);
+			     (char *) opc);
           immed = 0;
 	}
       else
@@ -1677,7 +2041,57 @@ md_assemble (char * str)
           output = frag_more (isize);
           immed = exp.X_add_number;
         }
-
+      if (streq (name, "brealid") || streq (name, "breaid") || streq (name, "breai"))
+	{
+	  temp = ((long long)immed) & 0xFFFFFFFFFFFF8000;
+	  if (temp != 0 && temp != 0xFFFFFFFFFFFF8000)
+	    {
+	      /* Needs an immediate inst.  */
+	   if (((long long)immed) > (long long)-549755813888 && ((long long)immed) < (long long)549755813887)
+	    {
+	    opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      if (opcode1 == NULL)
+		{
+		  as_bad (_("unknown opcode \"%s\""), "imml");
+		  return;
+		}
+	      inst1 = opcode1->bit_sequence;
+	      inst1 |= ((immed & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
+	    }
+	  else {
+	   opcode2 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	   opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      if (opcode1 == NULL || opcode2 == NULL)
+		{
+		  as_bad (_("unknown opcode \"%s\""), "imml");
+		  return;
+		}
+	      inst1 = opcode2->bit_sequence;
+	      inst1 |= ((immed & 0xFFFFFF0000000000L) >> 40) & IMML_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
+	      inst1 = opcode1->bit_sequence;
+	      inst1 |= ((immed & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
+	  }
+	  }
+	  inst |= (reg1 << RD_LOW) & RD_MASK;
+	  inst |= (immed << IMM_LOW) & IMM_MASK;
+	 }
+       else
+	 {
       temp = immed & 0xFFFF8000;
       if ((temp != 0) && (temp != 0xFFFF8000))
 	{
@@ -1700,6 +2114,7 @@ md_assemble (char * str)
 
       inst |= (reg1 << RD_LOW) & RD_MASK;
       inst |= (immed << IMM_LOW) & IMM_MASK;
+	}
       break;
 
     case INST_TYPE_R2:
@@ -1722,12 +2137,18 @@ md_assemble (char * str)
     case INST_TYPE_IMM:
       if (streq (name, "imm"))
         as_fatal (_("An IMM instruction should not be present in the .s file"));
-
-      op_end = parse_imm (op_end + 1, & exp, MIN_IMM, MAX_IMM);
+      if (microblaze_arch_size == 64)
+	op_end = parse_imml (op_end + 1, & exp, MIN_IMML, MAX_IMML);
+      else
+	op_end = parse_imm (op_end + 1, & exp, MIN_IMM, MAX_IMM);
 
       if (exp.X_op != O_constant)
 	{
-          char *opc = NULL;
+	  const char *opc;
+	  if (microblaze_arch_size == 64 && (streq (name, "brai") || streq (name, "braid")))
+	    opc = str_microblaze_64;
+	  else
+	    opc = NULL;
           relax_substateT subtype;
 
 	  if (exp.X_md != 0)
@@ -1741,7 +2162,7 @@ md_assemble (char * str)
 			     subtype,   /* PC-relative or not.  */
 			     exp.X_add_symbol,
 			     exp.X_add_number,
-			     opc);
+			     (char *) opc);
           immed = 0;
         }
       else
@@ -1750,27 +2171,79 @@ md_assemble (char * str)
           immed = exp.X_add_number;
         }
 
-
-      temp = immed & 0xFFFF8000;
-      if ((temp != 0) && (temp != 0xFFFF8000))
+      if (microblaze_arch_size == 64 && (streq (name, "breai") ||
+	    streq (name, "breaid") ||
+	    streq (name, "brai") || streq (name, "braid")))
 	{
-          /* Needs an immediate inst.  */
-          opcode1 = str_hash_find (opcode_hash_control, "imm");
-          if (opcode1 == NULL)
-            {
-              as_bad (_("unknown opcode \"%s\""), "imm");
-              return;
+	  temp = ((long long) immed) & 0xFFFFFFFFFFFF8000;
+	  if (temp != 0 && temp != 0xFFFFFFFFFFFF8000 && temp != 0x8000)
+	    {
+	      /* Needs an immediate inst.  */
+	  if (((long long)immed) > (long long)-549755813888 && ((long long)immed) < (long long)549755813887)
+	    {
+	      opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      if (opcode1 == NULL)
+		{
+		  as_bad (_("unknown opcode \"%s\""), "imml");
+		  return;
+		}
+	      inst1 = opcode1->bit_sequence;
+	      inst1 |= ((immed & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
+	    }
+	  else
+	    {
+	      opcode2 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      if (opcode1 == NULL || opcode2 == NULL)
+		{
+		  as_bad (_("unknown opcode \"%s\""), "imml");
+		  return;
+		}
+	      inst1 = opcode2->bit_sequence;
+	      inst1 |= ((immed & 0xFFFFFF0000000000L) >> 40) & IMML_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
+	      inst1 = opcode1->bit_sequence;
+	      inst1 |= ((immed & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
+	    }
+	  }
+	  inst |= (immed << IMM_LOW) & IMM_MASK;
+	}
+      else
+	{
+	  temp = immed & 0xFFFF8000;
+	  if ((temp != 0) && (temp != 0xFFFF8000))
+	    {
+	      /* Needs an immediate inst.  */
+	      opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imm");
+	      if (opcode1 == NULL)
+		{
+		  as_bad (_("unknown opcode \"%s\""), "imm");
+		  return;
+		}
+	      inst1 = opcode1->bit_sequence;
+	      inst1 |= ((immed & 0xFFFF0000) >> 16) & IMM_MASK;
+	      output[0] = INST_BYTE0 (inst1);
+	      output[1] = INST_BYTE1 (inst1);
+	      output[2] = INST_BYTE2 (inst1);
+	      output[3] = INST_BYTE3 (inst1);
+	      output = frag_more (isize);
             }
-
-          inst1 = opcode1->bit_sequence;
-          inst1 |= ((immed & 0xFFFF0000) >> 16) & IMM_MASK;
-          output[0] = INST_BYTE0 (inst1);
-          output[1] = INST_BYTE1 (inst1);
-          output[2] = INST_BYTE2 (inst1);
-          output[3] = INST_BYTE3 (inst1);
-          output = frag_more (isize);
-        }
-      inst |= (immed << IMM_LOW) & IMM_MASK;
+	  inst |= (immed << IMM_LOW) & IMM_MASK;
+	}
       break;
 
     case INST_TYPE_NONE:
@@ -1795,6 +2268,191 @@ md_assemble (char * str)
       inst |= (immed << IMM_MBAR);
       break;
 
+    /* For 64-bit instructions */
+    case INST_TYPE_RD_R1_IMML:
+      if (strcmp (op_end, ""))
+	op_end = parse_reg (op_end + 1, &reg1);  /* Get rd.  */
+      else
+	{
+	  as_fatal (_("Error in statement syntax"));
+	  reg1 = 0;
+	}
+      if (strcmp (op_end, ""))
+	op_end = parse_reg (op_end + 1, &reg2);  /* Get r1.  */
+      else
+	{
+	  as_fatal (_("Error in statement syntax"));
+	  reg2 = 0;
+	}
+      if (strcmp (op_end, ""))
+	op_end = parse_imml (op_end + 1, & exp, MIN_IMML, MAX_IMML);
+      else
+	as_fatal (_("Error in statement syntax"));
+
+      /* Check for spl registers.  */
+      if (check_spl_reg (& reg1))
+	as_fatal (_("Cannot use special register with this instruction"));
+      if (check_spl_reg (& reg2))
+	as_fatal (_("Cannot use special register with this instruction"));
+
+      if (exp.X_op != O_constant)
+	{
+	  const char *opc = str_microblaze_64;
+	  relax_substateT subtype;
+
+	  if (exp.X_md != 0)
+	    subtype = get_imm_otype(exp.X_md);
+	  else
+	    subtype = opcode->inst_offset_type;
+
+	  output = frag_var (rs_machine_dependent,
+			     isize * 2, /* maxm of 2 words.  */
+			     isize * 2, /* minm of 2 words.  */
+			     subtype,   /* PC-relative or not.  */
+			     exp.X_add_symbol,
+			     exp.X_add_number,
+			     (char *) opc);
+	  immedl = 0L;
+	}
+      else
+	{
+	  output = frag_more (isize);
+	  immedl = exp.X_add_number;
+	  if (((long long)immedl) > (long long)-549755813888 && ((long long)immedl) < (long long)549755813887)
+	    {
+	     opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	     if (opcode1 == NULL)
+	     {
+	      as_bad (_("unknown opcode \"%s\""), "imml");
+	      return;
+	     }
+	  inst1 = opcode1->bit_sequence;
+	  inst1 |= ((immedl & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	  output[0] = INST_BYTE0 (inst1);
+	  output[1] = INST_BYTE1 (inst1);
+	  output[2] = INST_BYTE2 (inst1);
+	  output[3] = INST_BYTE3 (inst1);
+	  output = frag_more (isize);
+	}
+	   else {
+	   opcode2 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	   opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+       if (opcode2 == NULL || opcode1 == NULL)
+	    {
+	      as_bad (_("unknown opcode \"%s\""), "imml");
+	      return;
+	    }
+	  inst1 = opcode2->bit_sequence;
+	  inst1 |= ((immedl & 0xFFFFFF0000000000L) >> 40) & IMML_MASK;
+	  output[0] = INST_BYTE0 (inst1);
+	  output[1] = INST_BYTE1 (inst1);
+	  output[2] = INST_BYTE2 (inst1);
+	  output[3] = INST_BYTE3 (inst1);
+	  output = frag_more (isize);
+	  inst1 = opcode1->bit_sequence;
+	  inst1 |= ((immedl & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	  output[0] = INST_BYTE0 (inst1);
+	  output[1] = INST_BYTE1 (inst1);
+	  output[2] = INST_BYTE2 (inst1);
+	  output[3] = INST_BYTE3 (inst1);
+	  output = frag_more (isize);
+	}
+	}
+
+      inst |= (reg1 << RD_LOW) & RD_MASK;
+      inst |= (reg2 << RA_LOW) & RA_MASK;
+      inst |= (immedl << IMM_LOW) & IMM_MASK;
+      break;
+
+    case INST_TYPE_R1_IMML:
+      if (strcmp (op_end, ""))
+	op_end = parse_reg (op_end + 1, &reg1);  /* Get r1.  */
+      else
+	{
+	  as_fatal (_("Error in statement syntax"));
+	  reg1 = 0;
+	}
+      if (strcmp (op_end, ""))
+	op_end = parse_imml (op_end + 1, & exp, MIN_IMML, MAX_IMML);
+      else
+	as_fatal (_("Error in statement syntax"));
+
+      /* Check for spl registers.  */
+      if (check_spl_reg (&reg1))
+	as_fatal (_("Cannot use special register with this instruction"));
+
+      if (exp.X_op != O_constant)
+	{
+	  const char *opc = str_microblaze_64;
+	  relax_substateT subtype;
+
+	  if (exp.X_md != 0)
+	    subtype = get_imm_otype(exp.X_md);
+	  else
+	    subtype = opcode->inst_offset_type;
+
+	  output = frag_var (rs_machine_dependent,
+			     isize * 2, /* maxm of 2 words.  */
+			     isize * 2, /* minm of 2 words.  */
+			     subtype,   /* PC-relative or not.  */
+			     exp.X_add_symbol,
+			     exp.X_add_number,
+			     (char *) opc);
+	  immedl = 0L;
+	}
+      else
+	{
+	  output = frag_more (isize);
+	  immedl = exp.X_add_number;
+	   if (((long long)immedl) > (long long)-549755813888 && ((long long)immedl) < (long long)549755813887)
+	     {
+	    opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	   if (opcode1 == NULL)
+	    {
+	      as_bad (_("unknown opcode \"%s\""), "imml");
+	      return;
+	    }
+	  inst1 = opcode1->bit_sequence;
+	  inst1 |= ((immedl & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	  output[0] = INST_BYTE0 (inst1);
+	  output[1] = INST_BYTE1 (inst1);
+	  output[2] = INST_BYTE2 (inst1);
+	  output[3] = INST_BYTE3 (inst1);
+	  output = frag_more (isize);
+	  }
+	   else {
+	   opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	   opcode2 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+       if (opcode2 == NULL || opcode1 == NULL)
+	    {
+	      as_bad (_("unknown opcode \"%s\""), "imml");
+	      return;
+	    }
+	  inst1 = opcode2->bit_sequence;
+	  inst1 |= ((immedl & 0xFFFFFF0000000000L) >> 40) & IMML_MASK;
+	  output[0] = INST_BYTE0 (inst1);
+	  output[1] = INST_BYTE1 (inst1);
+	  output[2] = INST_BYTE2 (inst1);
+	  output[3] = INST_BYTE3 (inst1);
+	  output = frag_more (isize);
+	  inst1 = opcode1->bit_sequence;
+	  inst1 |= ((immedl & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	  output[0] = INST_BYTE0 (inst1);
+	  output[1] = INST_BYTE1 (inst1);
+	  output[2] = INST_BYTE2 (inst1);
+	  output[3] = INST_BYTE3 (inst1);
+	  output = frag_more (isize);
+	}
+	}
+
+      inst |= (reg1 << RA_LOW) & RA_MASK;
+      inst |= (immedl << IMM_LOW) & IMM_MASK;
+      break;
+
+    case INST_TYPE_IMML:
+      as_fatal (_("An IMML instruction should not be present in the .s file"));
+      break;
+
     default:
       as_fatal (_("unimplemented opcode \"%s\""), name);
     }
@@ -1899,6 +2557,7 @@ const struct option md_longopts[] =
   {"EL", no_argument, NULL, OPTION_EL},
   {"mlittle-endian", no_argument, NULL, OPTION_EL},
   {"mbig-endian", no_argument, NULL, OPTION_EB},
+  {"m64", no_argument, NULL, OPTION_M64},
   { NULL,          no_argument, NULL, 0}
 };
 
@@ -1943,10 +2602,26 @@ md_convert_frag (bfd * abfd ATTRIBUTE_UNUSED,
       fragP->fr_fix += INST_WORD_SIZE * 2;
       fragP->fr_var = 0;
       break;
+    case DEFINED_64_PC_OFFSET:
+      fix_new (fragP, fragP->fr_fix, INST_WORD_SIZE, fragP->fr_symbol,
+		 fragP->fr_offset, TRUE, BFD_RELOC_MICROBLAZE_64_PCREL);
+      fragP->fr_fix += INST_WORD_SIZE * 2;
+      fragP->fr_var = 0;
+      break;
+    case DEFINED_64_OFFSET:
+      if (fragP->fr_symbol == GOT_symbol)
+	fix_new (fragP, fragP->fr_fix, INST_WORD_SIZE, fragP->fr_symbol,
+		 fragP->fr_offset, false, BFD_RELOC_MICROBLAZE_64_GPC);
+      else
+	fix_new (fragP, fragP->fr_fix, INST_WORD_SIZE, fragP->fr_symbol,
+		 fragP->fr_offset, false, BFD_RELOC_MICROBLAZE_64);
+      fragP->fr_fix += INST_WORD_SIZE * 2;
+      fragP->fr_var = 0;
+      break;
     case DEFINED_ABS_SEGMENT:
       if (fragP->fr_symbol == GOT_symbol)
         fix_new (fragP, fragP->fr_fix, INST_WORD_SIZE * 2, fragP->fr_symbol,
-	         fragP->fr_offset, true, BFD_RELOC_MICROBLAZE_64_GOTPC);
+		 fragP->fr_offset, true, BFD_RELOC_MICROBLAZE_64_GOTPC);
       else
         fix_new (fragP, fragP->fr_fix, INST_WORD_SIZE * 2, fragP->fr_symbol,
 	         fragP->fr_offset, false, BFD_RELOC_64);
@@ -2046,8 +2721,8 @@ md_apply_fix (fixS *   fixP,
   /* Note: use offsetT because it is signed, valueT is unsigned.  */
   offsetT      val  = *valp;
   int          i;
-  struct op_code_struct * opcode1;
-  unsigned long inst1;
+  struct op_code_struct * opcode1, * opcode2;
+  unsigned long inst1,inst2;
 
   symname = fixP->fx_addsy ? S_GET_NAME (fixP->fx_addsy) : _("<unknown>");
 
@@ -2148,18 +2823,74 @@ md_apply_fix (fixS *   fixP,
     case BFD_RELOC_RVA:
     case BFD_RELOC_32_PCREL:
     case BFD_RELOC_MICROBLAZE_32_SYM_OP_SYM:
+      /* Don't do anything if the symbol is not defined.  */
+      if (fixP->fx_addsy == NULL || S_IS_DEFINED (fixP->fx_addsy))
+	{
+	  if ((fixP->fx_r_type == BFD_RELOC_RVA) && (microblaze_arch_size == 64))
+	    {
+	      if (target_big_endian)
+		{
+		  buf[0] |= ((val >> 56) & 0xff);
+		  buf[1] |= ((val >> 48) & 0xff);
+		  buf[2] |= ((val >> 40) & 0xff);
+		  buf[3] |= ((val >> 32) & 0xff);
+		  buf[4] |= ((val >> 24) & 0xff);
+		  buf[5] |= ((val >> 16) & 0xff);
+		  buf[6] |= ((val >> 8) & 0xff);
+		  buf[7] |= (val & 0xff);
+		}
+	      else
+		{
+		  buf[7] |= ((val >> 56) & 0xff);
+		  buf[6] |= ((val >> 48) & 0xff);
+		  buf[5] |= ((val >> 40) & 0xff);
+		  buf[4] |= ((val >> 32) & 0xff);
+		  buf[3] |= ((val >> 24) & 0xff);
+		  buf[2] |= ((val >> 16) & 0xff);
+		  buf[1] |= ((val >> 8) & 0xff);
+		  buf[0] |= (val & 0xff);
+		}
+	    }
+	  else {
+	    if (target_big_endian)
+	      {
+		buf[0] |= ((val >> 24) & 0xff);
+		buf[1] |= ((val >> 16) & 0xff);
+		buf[2] |= ((val >> 8) & 0xff);
+		buf[3] |= (val & 0xff);
+	      }
+	    else
+	      {
+		buf[3] |= ((val >> 24) & 0xff);
+		buf[2] |= ((val >> 16) & 0xff);
+		buf[1] |= ((val >> 8) & 0xff);
+		buf[0] |= (val & 0xff);
+	      }
+	  }
+	}
+      break;
+
+    case BFD_RELOC_MICROBLAZE_EA64:
       /* Don't do anything if the symbol is not defined.  */
       if (fixP->fx_addsy == NULL || S_IS_DEFINED (fixP->fx_addsy))
 	{
 	  if (target_big_endian)
 	    {
-	      buf[0] |= ((val >> 24) & 0xff);
-	      buf[1] |= ((val >> 16) & 0xff);
-	      buf[2] |= ((val >> 8) & 0xff);
-	      buf[3] |= (val & 0xff);
+	      buf[0] |= ((val >> 56) & 0xff);
+	      buf[1] |= ((val >> 48) & 0xff);
+	      buf[2] |= ((val >> 40) & 0xff);
+	      buf[3] |= ((val >> 32) & 0xff);
+	      buf[4] |= ((val >> 24) & 0xff);
+	      buf[5] |= ((val >> 16) & 0xff);
+	      buf[6] |= ((val >> 8) & 0xff);
+	      buf[7] |= (val & 0xff);
 	    }
 	  else
 	    {
+	      buf[7] |= ((val >> 56) & 0xff);
+	      buf[6] |= ((val >> 48) & 0xff);
+	      buf[5] |= ((val >> 40) & 0xff);
+	      buf[4] |= ((val >> 32) & 0xff);
 	      buf[3] |= ((val >> 24) & 0xff);
 	      buf[2] |= ((val >> 16) & 0xff);
 	      buf[1] |= ((val >> 8) & 0xff);
@@ -2170,26 +2901,98 @@ md_apply_fix (fixS *   fixP,
     case BFD_RELOC_64_PCREL:
     case BFD_RELOC_64:
     case BFD_RELOC_MICROBLAZE_64_TEXTREL:
-      /* Add an imm instruction.  First save the current instruction.  */
-      for (i = 0; i < INST_WORD_SIZE; i++)
-	buf[i + INST_WORD_SIZE] = buf[i];
+    case BFD_RELOC_MICROBLAZE_64:
+    case BFD_RELOC_MICROBLAZE_64_PCREL:
 
-      /* Generate the imm instruction.  */
-      opcode1 = str_hash_find (opcode_hash_control, "imm");
-      if (opcode1 == NULL)
+      if (fixP->fx_r_type == BFD_RELOC_MICROBLAZE_64
+	    || fixP->fx_r_type == BFD_RELOC_MICROBLAZE_64_PCREL || (fixP->fx_r_type == BFD_RELOC_64_PCREL && microblaze_arch_size == 64))
 	{
-	  as_bad (_("unknown opcode \"%s\""), "imm");
-	  return;
+	  /* Generate the imm instruction.  */
+	  if (((long long)val) > (long long)-549755813888 && ((long long)val) < (long long)549755813887)
+	    {
+	      /* Add an imm instruction.  First save the current instruction.  */
+	      for (i = 0; i < INST_WORD_SIZE; i++)
+		buf[i + INST_WORD_SIZE] = buf[i];
+	      opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      if (opcode1 == NULL)
+		{
+		  as_bad (_("unknown opcode \"%s\""), "imml");
+		  return;
+		}
+	      inst1 = opcode1->bit_sequence;
+	      if (fixP->fx_addsy == NULL || S_IS_DEFINED (fixP->fx_addsy))
+		inst1 |= ((val & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	      if (fixP->fx_r_type == BFD_RELOC_MICROBLAZE_64)
+		fixP->fx_r_type = BFD_RELOC_64;
+	      if (fixP->fx_r_type == BFD_RELOC_MICROBLAZE_64_PCREL)
+		fixP->fx_r_type = BFD_RELOC_64_PCREL;
+
+	      buf[0] = INST_BYTE0 (inst1);
+	      buf[1] = INST_BYTE1 (inst1);
+	      buf[2] = INST_BYTE2 (inst1);
+	      buf[3] = INST_BYTE3 (inst1);
+	    }
+	  else
+	    {
+	      /* Add an imm instruction.  First save the current instruction.  */
+	      for (i = 0; i < INST_WORD_SIZE; i++)
+		buf[i + INST_WORD_SIZE + 4] = buf[i];
+
+	      opcode2 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      if (opcode1 == NULL || opcode2 ==NULL)
+		{
+		  as_bad (_("unknown opcode \"%s\""), "imml");
+		  return;
+		}
+	      inst1 = opcode2->bit_sequence;
+	      if (fixP->fx_addsy == NULL || S_IS_DEFINED (fixP->fx_addsy))
+		inst1 |= ((val & 0x000000FFFFFF0000L) >> 40) & IMML_MASK;
+	      if (fixP->fx_r_type == BFD_RELOC_MICROBLAZE_64)
+		fixP->fx_r_type = BFD_RELOC_64;
+	      if (fixP->fx_r_type == BFD_RELOC_MICROBLAZE_64_PCREL)
+		fixP->fx_r_type = BFD_RELOC_64_PCREL;
+	      inst2 = opcode1->bit_sequence;
+	      if (fixP->fx_addsy == NULL || S_IS_DEFINED (fixP->fx_addsy))
+		inst1 |= ((val & 0x000000FFFFFF0000L) >> 16) & IMML_MASK;
+	    if (fixP->fx_r_type == BFD_RELOC_MICROBLAZE_64)
+	      fixP->fx_r_type = BFD_RELOC_64;
+	    if (fixP->fx_r_type == BFD_RELOC_MICROBLAZE_64_PCREL)
+	      fixP->fx_r_type = BFD_RELOC_64_PCREL;
+
+	    buf[0] = INST_BYTE0 (inst1);
+	    buf[1] = INST_BYTE1 (inst1);
+	    buf[2] = INST_BYTE2 (inst1);
+	    buf[3] = INST_BYTE3 (inst1);
+	    buf[4] = INST_BYTE0 (inst2);
+	    buf[5] = INST_BYTE1 (inst2);
+	    buf[6] = INST_BYTE2 (inst2);
+	    buf[7] = INST_BYTE3 (inst2);
+	  }
 	}
+      else
+	{
+	  /* Add an imm instruction.  First save the current instruction.  */
+	  for (i = 0; i < INST_WORD_SIZE; i++)
+	    buf[i + INST_WORD_SIZE] = buf[i];
 
-      inst1 = opcode1->bit_sequence;
-      if (fixP->fx_addsy == NULL || S_IS_DEFINED (fixP->fx_addsy))
-	inst1 |= ((val & 0xFFFF0000) >> 16) & IMM_MASK;
+	  /* Generate the imm instruction.  */
+	  opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imm");
+	  if (opcode1 == NULL)
+	    {
+	      as_bad (_("unknown opcode \"%s\""), "imm");
+	      return;
+	    }
 
-      buf[0] = INST_BYTE0 (inst1);
-      buf[1] = INST_BYTE1 (inst1);
-      buf[2] = INST_BYTE2 (inst1);
-      buf[3] = INST_BYTE3 (inst1);
+	  inst1 = opcode1->bit_sequence;
+	  if (fixP->fx_addsy == NULL || S_IS_DEFINED (fixP->fx_addsy))
+	    inst1 |= ((val & 0xFFFF0000) >> 16) & IMM_MASK;
+
+	  buf[0] = INST_BYTE0 (inst1);
+	  buf[1] = INST_BYTE1 (inst1);
+	  buf[2] = INST_BYTE2 (inst1);
+	  buf[3] = INST_BYTE3 (inst1);
+	}
 
       /* Add the value only if the symbol is defined.  */
       if (fixP->fx_addsy == NULL || S_IS_DEFINED (fixP->fx_addsy))
@@ -2214,6 +3017,7 @@ md_apply_fix (fixS *   fixP,
       /* Fall through.  */
 
     case BFD_RELOC_MICROBLAZE_64_GOTPC:
+    case BFD_RELOC_MICROBLAZE_64_GPC:
     case BFD_RELOC_MICROBLAZE_64_GOT:
     case BFD_RELOC_MICROBLAZE_64_PLT:
     case BFD_RELOC_MICROBLAZE_64_GOTOFF:
@@ -2221,12 +3025,42 @@ md_apply_fix (fixS *   fixP,
       /* Add an imm instruction.  First save the current instruction.  */
       for (i = 0; i < INST_WORD_SIZE; i++)
 	buf[i + INST_WORD_SIZE] = buf[i];
-
       /* Generate the imm instruction.  */
-      opcode1 = str_hash_find (opcode_hash_control, "imm");
+      if (fixP->fx_r_type == BFD_RELOC_MICROBLAZE_64_GPC)
+	{
+	  if (((long long)val) > (long long)-549755813888 && ((long long)val) < (long long)549755813887)
+	    {
+	      for (i = 0; i < INST_WORD_SIZE; i++)
+		buf[i + INST_WORD_SIZE] = buf[i];
+	      opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	    }
+	  else
+	    {
+	      for (i = 0; i < INST_WORD_SIZE; i++)
+		buf[i + INST_WORD_SIZE + 4] = buf[i];
+	      opcode2 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imml");
+	      inst2 = opcode2->bit_sequence;
+
+	      /* We can fixup call to a defined non-global address
+	       * within the same section only.  */
+	      buf[4] = INST_BYTE0 (inst2);
+	      buf[5] = INST_BYTE1 (inst2);
+	      buf[6] = INST_BYTE2 (inst2);
+	      buf[7] = INST_BYTE3 (inst2);
+	   }
+	}
+      else
+	opcode1 = (struct op_code_struct *) str_hash_find (opcode_hash_control, "imm");
       if (opcode1 == NULL)
 	{
-	  as_bad (_("unknown opcode \"%s\""), "imm");
+	  for (i = 0; i < INST_WORD_SIZE; i++)
+	    buf[i + INST_WORD_SIZE] = buf[i];
+
+	  if (fixP->fx_r_type == BFD_RELOC_MICROBLAZE_64_GPC)
+	    as_bad (_("unknown opcode \"%s\""), "imml");
+	  else
+	    as_bad (_("unknown opcode \"%s\""), "imm");
 	  return;
 	}
 
@@ -2252,6 +3086,8 @@ md_apply_fix (fixS *   fixP,
 	fixP->fx_r_type = BFD_RELOC_MICROBLAZE_64_NONE;
       else if (fixP->fx_r_type == BFD_RELOC_32)
 	fixP->fx_r_type = BFD_RELOC_MICROBLAZE_32_NONE;
+      else if(fixP->fx_r_type == BFD_RELOC_MICROBLAZE_EA64)
+	fixP->fx_r_type = BFD_RELOC_MICROBLAZE_EA64;
       else
 	fixP->fx_r_type = BFD_RELOC_NONE;
       fixP->fx_addsy = section_symbol (absolute_section);
@@ -2292,12 +3128,30 @@ md_estimate_size_before_relax (fragS * fragP,
           as_bad (_("Absolute PC-relative value in relaxation code.  Assembler error....."));
           abort ();
         }
-      else if (S_GET_SEGMENT (fragP->fr_symbol) == segment_type &&
-               !S_IS_WEAK (fragP->fr_symbol))
+      else if (S_GET_SEGMENT (fragP->fr_symbol) == segment_type
+	       && !S_IS_WEAK (fragP->fr_symbol))
         {
-          fragP->fr_subtype = DEFINED_PC_OFFSET;
-          /* Don't know now whether we need an imm instruction.  */
-          fragP->fr_var = INST_WORD_SIZE;
+	   if (fragP->fr_opcode != NULL) {
+	     if(streq (fragP->fr_opcode, str_microblaze_64))
+	     {
+	       /* Used as an absolute value.  */
+	      fragP->fr_subtype = DEFINED_64_PC_OFFSET;
+	       /* Variable part does not change.  */
+	      fragP->fr_var = INST_WORD_SIZE*2;
+	     }
+	   else
+	     {
+	       fragP->fr_subtype = DEFINED_PC_OFFSET;
+	       /* Don't know now whether we need an imm instruction.  */
+	       fragP->fr_var = INST_WORD_SIZE;
+	     }
+	   }
+	   else
+	     {
+	       fragP->fr_subtype = DEFINED_PC_OFFSET;
+	       /* Don't know now whether we need an imm instruction.  */
+	       fragP->fr_var = INST_WORD_SIZE;
+	     }
         }
       else if (S_IS_DEFINED (fragP->fr_symbol)
 	       && (((S_GET_SEGMENT (fragP->fr_symbol))->flags & SEC_CODE) == 0))
@@ -2310,7 +3164,14 @@ md_estimate_size_before_relax (fragS * fragP,
         }
       else
 	{
-	  fragP->fr_subtype = UNDEFINED_PC_OFFSET;
+	  if (fragP->fr_opcode != NULL) {
+	    if (streq (fragP->fr_opcode, str_microblaze_64))
+	      fragP->fr_subtype = DEFINED_64_PC_OFFSET;
+	    else
+	      fragP->fr_subtype = UNDEFINED_PC_OFFSET;
+	    }
+	  else
+	    fragP->fr_subtype = UNDEFINED_PC_OFFSET;
 	  fragP->fr_var = INST_WORD_SIZE*2;
 	}
       break;
@@ -2328,6 +3189,13 @@ md_estimate_size_before_relax (fragS * fragP,
 	      /* Variable part does not change.  */
 	      fragP->fr_var = INST_WORD_SIZE*2;
 	    }
+	  else if (streq (fragP->fr_opcode, str_microblaze_64))
+	    {
+	      /* Used as an absolute value.  */
+	      fragP->fr_subtype = DEFINED_64_OFFSET;
+	      /* Variable part does not change.  */
+	      fragP->fr_var = INST_WORD_SIZE;
+	    }
 	  else if (streq (fragP->fr_opcode, str_microblaze_ro_anchor))
 	    {
               /* It is accessed using the small data read only anchor.  */
@@ -2401,6 +3269,8 @@ md_estimate_size_before_relax (fragS * fragP,
     case TLSLD_OFFSET:
     case TLSTPREL_OFFSET:
     case TLSDTPREL_OFFSET:
+    case DEFINED_64_OFFSET:
+    case DEFINED_64_PC_OFFSET:
       fragP->fr_var = INST_WORD_SIZE*2;
       break;
     case DEFINED_RO_SEGMENT:
@@ -2454,7 +3324,7 @@ md_pcrel_from_section (fixS * fixp, segT sec ATTRIBUTE_UNUSED)
   else
     {
       /* The case where we are going to resolve things... */
-      if (fixp->fx_r_type == BFD_RELOC_64_PCREL)
+      if (fixp->fx_r_type == BFD_RELOC_64_PCREL ||fixp->fx_r_type == BFD_RELOC_MICROBLAZE_64_PCREL)
         return  fixp->fx_where + fixp->fx_frag->fr_address + INST_WORD_SIZE;
       else
         return  fixp->fx_where + fixp->fx_frag->fr_address;
@@ -2487,6 +3357,10 @@ tc_gen_reloc (asection * section ATTRIBUTE_UNUSED, fixS * fixp)
     case BFD_RELOC_MICROBLAZE_32_RWSDA:
     case BFD_RELOC_MICROBLAZE_32_SYM_OP_SYM:
     case BFD_RELOC_MICROBLAZE_64_GOTPC:
+    case BFD_RELOC_MICROBLAZE_64_GPC:
+    case BFD_RELOC_MICROBLAZE_EA64:
+    case BFD_RELOC_MICROBLAZE_64:
+    case BFD_RELOC_MICROBLAZE_64_PCREL:
     case BFD_RELOC_MICROBLAZE_64_GOT:
     case BFD_RELOC_MICROBLAZE_64_PLT:
     case BFD_RELOC_MICROBLAZE_64_GOTOFF:
@@ -2547,6 +3421,18 @@ tc_gen_reloc (asection * section ATTRIBUTE_UNUSED, fixS * fixp)
   return rel;
 }
 
+/* Called by TARGET_FORMAT.  */
+const char *
+microblaze_target_format (void)
+{
+
+  if (microblaze_arch_size == 64)
+    return "elf64-microblazeel";
+  else
+      return target_big_endian ? "elf32-microblaze" : "elf32-microblazeel";
+}
+
+
 int
 md_parse_option (int c, const char * arg ATTRIBUTE_UNUSED)
 {
@@ -2560,6 +3446,10 @@ md_parse_option (int c, const char * arg ATTRIBUTE_UNUSED)
     case OPTION_LITTLE:
       target_big_endian = 0;
       break;
+    case OPTION_M64:
+      //if (arg != NULL && strcmp (arg, "64") == 0)
+      microblaze_arch_size = 64;
+      break;
     default:
       return 0;
     }
@@ -2575,6 +3465,7 @@ md_show_usage (FILE * stream ATTRIBUTE_UNUSED)
   fprintf (stream, _(" MicroBlaze specific assembler options:\n"));
   fprintf (stream, "  -%-23s%s\n", "mbig-endian", N_("assemble for a big endian cpu"));
   fprintf (stream, "  -%-23s%s\n", "mlittle-endian", N_("assemble for a little endian cpu"));      
+  fprintf (stream, "  -%-23s%s\n", "m64", N_("generate 64-bit elf"));
 }
 
 
@@ -2612,7 +3503,10 @@ cons_fix_new_microblaze (fragS * frag,
           r = BFD_RELOC_32;
           break;
         case 8:
-          r = BFD_RELOC_64;
+	  if (microblaze_arch_size == 64)
+	    r = BFD_RELOC_MICROBLAZE_EA64;
+	  else
+	    r = BFD_RELOC_64;
           break;
         default:
           as_bad (_("unsupported BFD relocation size %u"), size);
diff --git a/gas/config/tc-microblaze.h b/gas/config/tc-microblaze.h
index 1fe678d72e7..90a8277f71e 100644
--- a/gas/config/tc-microblaze.h
+++ b/gas/config/tc-microblaze.h
@@ -87,7 +87,9 @@ extern const struct relax_type md_relax_table[];
 
 #ifdef OBJ_ELF
 
-#define TARGET_FORMAT (target_big_endian ? "elf32-microblaze" : "elf32-microblazeel")
+#define TARGET_FORMAT microblaze_target_format()
+extern const char *microblaze_target_format (void);
+//#define TARGET_FORMAT (target_big_endian ? "elf32-microblaze" : "elf32-microblazeel")
 
 #define ELF_TC_SPECIAL_SECTIONS \
   { ".sdata",		SHT_PROGBITS,	SHF_ALLOC + SHF_WRITE }, \
diff --git a/gas/testsuite/gas/microblaze/allinsn64.d b/gas/testsuite/gas/microblaze/allinsn64.d
new file mode 100644
index 00000000000..46b686686f3
--- /dev/null
+++ b/gas/testsuite/gas/microblaze/allinsn64.d
@@ -0,0 +1,252 @@
+#as: -m64
+#objdump: -d
+
+.*: +file format .*
+
+Disassembly of section .text:
+
+0000000000000000 <addl>:
+   0:	00875900 	addl	r4, r7, r11
+
+0000000000000004 <addlc>:
+   4:	08875900 	addlc	r4, r7, r11
+
+0000000000000008 <addlk>:
+   8:	10875900 	addlk	r4, r7, r11
+
+000000000000000c <addlkc>:
+   c:	18875900 	addlkc	r4, r7, r11
+
+0000000000000010 <addli>:
+  10:	b2000000 	imml	0
+  14:	20871001 	addi	r4, r7, 4097
+
+0000000000000018 <z1addli>:
+  18:	b2000001 	imml	1
+  1c:	20870000 	addi	r4, r7, 0
+
+0000000000000020 <z2addli>:
+  20:	b2100000 	imml	1048576
+  24:	b2000000 	imml	0
+  28:	20870000 	addi	r4, r7, 0
+
+000000000000002c <addlic>:
+  2c:	b2000000 	imml	0
+  30:	28871001 	addic	r4, r7, 4097
+
+0000000000000034 <addlik>:
+  34:	b2000000 	imml	0
+  38:	30871001 	addik	r4, r7, 4097
+
+000000000000003c <addlikc>:
+  3c:	b2000000 	imml	0
+  40:	38871001 	addikc	r4, r7, 4097
+
+0000000000000044 <z3addli>:
+  44:	68801001 	addli	r4, 4097
+
+0000000000000048 <z1addlic>:
+  48:	68821001 	addlic	r4, 4097
+
+000000000000004c <z1addlik>:
+  4c:	68841001 	addlik	r4, 4097
+
+0000000000000050 <z1addlikc>:
+  50:	68861001 	addlikc	r4, 4097
+
+0000000000000054 <andl>:
+  54:	84875900 	andl	r4, r7, r11
+
+0000000000000058 <andli>:
+  58:	b2000000 	imml	0
+  5c:	a4871001 	andi	r4, r7, 4097
+
+0000000000000060 <z1andli>:
+  60:	68911001 	andli	r4, 4097
+
+0000000000000064 <andnl>:
+  64:	8c875900 	andnl	r4, r7, r11
+
+0000000000000068 <andnli>:
+  68:	b2000000 	imml	0
+  6c:	ac871001 	andni	r4, r7, 4097
+
+0000000000000070 <z1andnli>:
+  70:	68931001 	andnli	r4, 4097
+
+0000000000000074 <bslrl>:
+  74:	44875900 	bslrl	r4, r7, r11
+
+0000000000000078 <bslra>:
+  78:	44875b00 	bslra	r4, r7, r11
+
+000000000000007c <bslll>:
+  7c:	44875d00 	bslll	r4, r7, r11
+
+0000000000000080 <bslrli>:
+  80:	64872023 	bslrli	r4, r7, 35
+
+0000000000000084 <bslrai>:
+  84:	64872223 	bslrai	r4, r7, 35
+
+0000000000000088 <bsllli>:
+  88:	64872423 	bsllli	r4, r7, 35
+
+000000000000008c <bslefi>:
+  8c:	648764df 	bslefi	r4, r7, 19, 31
+
+0000000000000090 <bslifi>:
+  90:	6487ac5f 	bslifi	r4, r7, 19, 31
+
+0000000000000094 <cmpl>:
+  94:	14875901 	cmpl	r4, r7, r11
+
+0000000000000098 <cmplu>:
+  98:	14875903 	cmplu	r4, r7, r11
+
+000000000000009c <dadd>:
+  9c:	58875c00 	dadd	r4, r7, r11
+
+00000000000000a0 <drsub>:
+  a0:	58875c80 	drsub	r4, r7, r11
+
+00000000000000a4 <dmul>:
+  a4:	58875d00 	dmul	r4, r7, r11
+
+00000000000000a8 <ddiv>:
+  a8:	58875d80 	ddiv	r4, r7, r11
+
+00000000000000ac <dcmp.un>:
+  ac:	58875e00 	dcmp.un	r4, r7, r11
+
+00000000000000b0 <dcmp.lt>:
+  b0:	58875e10 	dcmp.lt	r4, r7, r11
+
+00000000000000b4 <dcmp.eq>:
+  b4:	58875e20 	dcmp.eq	r4, r7, r11
+
+00000000000000b8 <dcmp.le>:
+  b8:	58875e30 	dcmp.le	r4, r7, r11
+
+00000000000000bc <dcmp.gt>:
+  bc:	58875e40 	dcmp.gt	r4, r7, r11
+
+00000000000000c0 <dcmp.ne>:
+  c0:	58875e50 	dcmp.ne	r4, r7, r11
+
+00000000000000c4 <dcmp.ge>:
+  c4:	58875e60 	dcmp.ge	r4, r7, r11
+
+00000000000000c8 <dbl>:
+  c8:	58870680 	dbl	r4, r7
+
+00000000000000cc <dlong>:
+  cc:	58870700 	dlong	r4, r7
+
+00000000000000d0 <dsqrt>:
+  d0:	58870780 	dsqrt	r4, r7
+
+00000000000000d4 <ll>:
+  d4:	c8875900 	ll	r4, r7, r11
+
+00000000000000d8 <llr>:
+  d8:	c8875b00 	llr	r4, r7, r11
+
+00000000000000dc <lli>:
+  dc:	ec871000 	lli	r4, r7, 4096
+
+00000000000000e0 <orl>:
+  e0:	80875900 	orl	r4, r7, r11
+
+00000000000000e4 <orli>:
+  e4:	b2000000 	imml	0
+  e8:	a0871000 	ori	r4, r7, 4096
+
+00000000000000ec <z1orli>:
+  ec:	68901000 	orli	r4, 4096
+
+00000000000000f0 <pcmplbf>:
+  f0:	80875d00 	pcmplbf	r4, r7, r11
+
+00000000000000f4 <pcmpleq>:
+  f4:	88875d00 	pcmpleq	r4, r7, r11
+
+00000000000000f8 <pcmplne>:
+  f8:	8c875d00 	pcmplne	r4, r7, r11
+
+00000000000000fc <rsubl>:
+  fc:	04875900 	rsubl	r4, r7, r11
+
+0000000000000100 <rsublc>:
+ 100:	0c875900 	rsublc	r4, r7, r11
+
+0000000000000104 <rsublk>:
+ 104:	14875900 	rsublk	r4, r7, r11
+
+0000000000000108 <rsublkc>:
+ 108:	1c875900 	rsublkc	r4, r7, r11
+
+000000000000010c <rsubli>:
+ 10c:	b2000000 	imml	0
+ 110:	24871000 	rsubi	r4, r7, 4096
+
+0000000000000114 <rsublic>:
+ 114:	b2000000 	imml	0
+ 118:	2c871000 	rsubic	r4, r7, 4096
+
+000000000000011c <rsublik>:
+ 11c:	b2000000 	imml	0
+ 120:	34871000 	rsubik	r4, r7, 4096
+
+0000000000000124 <rsublikc>:
+ 124:	b2000000 	imml	0
+ 128:	3c871000 	rsubikc	r4, r7, 4096
+
+000000000000012c <z1rsubli>:
+ 12c:	68811000 	rsubli	r4, 4096
+
+0000000000000130 <z1rsublic>:
+ 130:	68831000 	rsublic	r4, 4096
+
+0000000000000134 <z1rsublik>:
+ 134:	68851000 	rsublik	r4, 4096
+
+0000000000000138 <z1rsublikc>:
+ 138:	68871000 	rsublikc	r4, 4096
+
+000000000000013c <sextl16>:
+ 13c:	90870161 	sextl16	r4, r7
+
+0000000000000140 <sextl32>:
+ 140:	90870162 	sextl32	r4, r7
+
+0000000000000144 <sextl8>:
+ 144:	90870160 	sextl8	r4, r7
+
+0000000000000148 <srla>:
+ 148:	90870101 	srla	r4, r7
+
+000000000000014c <srlc>:
+ 14c:	90870121 	srlc	r4, r7
+
+0000000000000150 <srll>:
+ 150:	90870141 	srll	r4, r7
+
+0000000000000154 <sl>:
+ 154:	d8875900 	sl	r4, r7, r11
+
+0000000000000158 <slr>:
+ 158:	d8875b00 	slr	r4, r7, r11
+
+000000000000015c <sli>:
+ 15c:	fc871000 	sli	r4, r7, 4096
+
+0000000000000160 <xorl>:
+ 160:	88875900 	xorl	r4, r7, r11
+
+0000000000000164 <xorli>:
+ 164:	b2000000 	imml	0
+ 168:	a8871000 	xori	r4, r7, 4096
+
+000000000000016c <z1xorli>:
+ 16c:	68921000 	xorli	r4, 4096
diff --git a/gas/testsuite/gas/microblaze/allinsn64.exp b/gas/testsuite/gas/microblaze/allinsn64.exp
new file mode 100644
index 00000000000..c2391d0d2ed
--- /dev/null
+++ b/gas/testsuite/gas/microblaze/allinsn64.exp
@@ -0,0 +1,21 @@
+# Copyright (C) 2012-2024 Free Software Foundation, Inc.
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street - Fifth Floor, Boston, MA 02110-1301, USA.
+
+# MicroBlaze test for special register.
+
+if [istarget microblaze*-*-*] {
+    run_dump_test "allinsn64"
+}
diff --git a/gas/testsuite/gas/microblaze/allinsn64.s b/gas/testsuite/gas/microblaze/allinsn64.s
new file mode 100644
index 00000000000..b9cbf0cd290
--- /dev/null
+++ b/gas/testsuite/gas/microblaze/allinsn64.s
@@ -0,0 +1,234 @@
+    .text
+footext:
+    .text
+	.global addl
+addl:
+	addl r4, r7, r11
+	.global addlc
+addlc:
+	addlc r4, r7, r11
+	.global addlk
+addlk:
+	addlk r4, r7, r11
+	.global addlkc
+addlkc:
+	addlkc r4, r7, r11
+	.global addli
+addli:
+	addli r4, r7, 4097
+	.global z1addli
+z1addli:
+	addli r4, r7, 0x0000000000010000
+	.global z2addli
+z2addli:
+	addli r4, r7, 0x1000000000000000
+	.global addlic
+addlic:
+	addlic r4, r7, 4097
+	.global addlik
+addlik:
+	addlik r4, r7, 4097
+	.global addlikc
+addlikc:
+	addlikc r4, r7, 4097
+	.global z3addli
+z3addli:
+	addli r4, 4097
+	.global z1addlic
+z1addlic:
+	addlic r4, 4097
+	.global z1addlik
+z1addlik:
+	addlik r4, 4097
+	.global z1addlikc
+z1addlikc:
+	addlikc r4, 4097
+	.global andl
+andl:
+	andl r4, r7, r11
+	.global andli
+andli:
+	andli r4, r7, 4097
+	.global z1andli
+z1andli:
+	andli r4, 4097
+	.global andnl
+andnl:
+	andnl r4, r7, r11
+	.global andnli
+andnli:
+	andnli r4,r7,4097
+	.global z1andnli
+z1andnli:
+	andnli r4,4097
+	.global bslrl
+bslrl:
+	bslrl r4,r7,r11
+	.global bslra
+bslra:
+	bslra r4,r7,r11
+	.global bslll
+bslll:
+	bslll r4,r7,r11
+	.global bslrli
+bslrli:
+	bslrli r4,r7,35
+	.global bslrai
+bslrai:
+	bslrai r4,r7,35
+	.global bsllli
+bsllli:
+	bsllli r4,r7,35
+	.global bslefi
+bslefi:
+	bslefi r4,r7,19,31
+	.global bslifi
+bslifi:
+	bslifi r4,r7,19,31
+	.global cmpl
+cmpl:
+	cmpl r4, r7, r11
+	.global cmplu
+cmplu:
+	cmplu r4, r7, r11
+	.global dadd
+dadd:
+	dadd r4, r7, r11
+	.global drsub
+drsub:
+	drsub r4, r7, r11
+	.global dmul
+dmul:
+	dmul r4, r7, r11
+	.global ddiv
+ddiv:
+	ddiv r4, r7, r11
+	.global dcmp.un
+dcmp.un:
+	dcmp.un r4,r7,r11
+	.global dcmp.lt
+dcmp.lt:
+	dcmp.lt r4,r7,r11
+	.global dcmp.eq
+dcmp.eq:
+	dcmp.eq r4,r7,r11
+	.global dcmp.le
+dcmp.le:
+	dcmp.le r4,r7,r11
+	.global dcmp.gt
+dcmp.gt:
+	dcmp.gt r4,r7,r11
+	.global dcmp.ne
+dcmp.ne:
+	dcmp.ne r4,r7,r11
+	.global dcmp.ge
+dcmp.ge:
+	dcmp.ge r4,r7,r11
+	.global dbl
+dbl:
+	dbl r4, r7
+	.global dlong
+dlong:
+	dlong r4, r7
+	.global dsqrt
+dsqrt:
+	dsqrt r4,r7
+	.global ll
+ll:
+	ll r4, r7, r11
+	.global llr
+llr:
+	llr r4, r7, r11
+	.global lli
+lli:
+	lli r4,r7,4096
+	.global orl
+orl:
+	orl r4, r7, r11
+	.global orli
+orli:
+	orli r4, r7, 4096
+	.global z1orli
+z1orli:
+	orli r4, 4096
+	.global pcmplbf
+pcmplbf:
+	pcmplbf r4, r7, r11
+	.global pcmpleq
+pcmpleq:
+	pcmpleq r4, r7, r11
+	.global pcmplne
+pcmplne:
+	pcmplne r4, r7, r11
+	.global rsubl
+rsubl:
+	rsubl r4, r7, r11
+	.global rsublc
+rsublc:
+	rsublc r4, r7, r11
+	.global rsublk
+rsublk:
+	rsublk r4, r7, r11
+	.global rsublkc
+rsublkc:
+	rsublkc r4, r7, r11
+	.global rsubli
+rsubli:
+	rsubli r4, r7, 4096
+	.global rsublic
+rsublic:
+	rsublic r4, r7, 4096
+	.global rsublik
+rsublik:
+	rsublik r4, r7, 4096
+	.global rsublikc
+rsublikc:
+	rsublikc r4, r7, 4096
+	.global z1rsubli
+z1rsubli:
+	rsubli r4, 4096
+	.global z1rsublic
+z1rsublic:
+	rsublic r4, 4096
+	.global z1rsublik
+z1rsublik:
+	rsublik r4, 4096
+	.global z1rsublikc
+z1rsublikc:
+	rsublikc r4, 4096
+	.global sextl16
+sextl16:
+	sextl16 r4, r7
+	.global sextl32
+sextl32:
+	sextl32 r4, r7
+	.global sextl8
+sextl8:
+	sextl8 r4, r7
+	.global srla
+srla:
+	srla r4, r7
+	.global srlc
+srlc:
+	srlc r4, r7
+	.global srll
+srll:
+	srll r4, r7
+	.global sl
+sl:
+	sl r4, r7, r11
+	.global slr
+slr:
+	slr r4, r7, r11
+	.global sli
+sli:
+	sli r4, r7, 4096
+	.global xorl
+xorl:
+	xorl r4, r7, r11
+	.global xorli
+xorli:
+	xorli r4, r7, 4096
+	.global z1xorli
+z1xorli:
+	xorli r4, 4096
diff --git a/gas/testsuite/gas/microblaze/brea.d b/gas/testsuite/gas/microblaze/brea.d
new file mode 100644
index 00000000000..b055554ec6d
--- /dev/null
+++ b/gas/testsuite/gas/microblaze/brea.d
@@ -0,0 +1,132 @@
+#as: -m64
+#objdump: -d
+
+.*: +file format .*
+
+Disassembly of section .text:
+
+0000000000000000 <beaeq>:
+   0:	9d043800 	beaeq	r4, r7
+
+0000000000000004 <bealeq>:
+   4:	9d043900 	bealeq	r4, r7
+
+0000000000000008 <beaeqd>:
+   8:	9f043800 	beaeqd	r4, r7
+
+000000000000000c <bealeqd>:
+   c:	9f043900 	bealeqd	r4, r7
+
+0000000000000010 <beage>:
+  10:	9da43800 	beage	r4, r7
+
+0000000000000014 <bealge>:
+  14:	9da43900 	bealge	r4, r7
+
+0000000000000018 <beaged>:
+  18:	9fa43800 	beaged	r4, r7
+
+000000000000001c <bealged>:
+  1c:	9fa43900 	bealged	r4, r7
+
+0000000000000020 <beagt>:
+  20:	9d843800 	beagt	r4, r7
+
+0000000000000024 <bealgt>:
+  24:	9d843900 	bealgt	r4, r7
+
+0000000000000028 <beagtd>:
+  28:	9f843800 	beagtd	r4, r7
+
+000000000000002c <bealgtd>:
+  2c:	9f843900 	bealgtd	r4, r7
+
+0000000000000030 <beale>:
+  30:	9d643800 	beale	r4, r7
+
+0000000000000034 <bealle>:
+  34:	9d643900 	bealle	r4, r7
+
+0000000000000038 <bealed>:
+  38:	9f643800 	bealed	r4, r7
+
+000000000000003c <bealled>:
+  3c:	9f643900 	bealled	r4, r7
+
+0000000000000040 <bealt>:
+  40:	9d443800 	bealt	r4, r7
+
+0000000000000044 <beallt>:
+  44:	9d443900 	beallt	r4, r7
+
+0000000000000048 <bealtd>:
+  48:	9f443800 	bealtd	r4, r7
+
+000000000000004c <bealltd>:
+  4c:	9f443900 	bealltd	r4, r7
+
+0000000000000050 <beane>:
+  50:	9d243800 	beane	r4, r7
+
+0000000000000054 <bealne>:
+  54:	9d243900 	bealne	r4, r7
+
+0000000000000058 <beaned>:
+  58:	9f243800 	beaned	r4, r7
+
+000000000000005c <bealned>:
+  5c:	9f243900 	bealned	r4, r7
+
+0000000000000060 <beaeqi>:
+  60:	bd041000 	beaeqi	r4, 4096		// 1060
+
+0000000000000064 <beaeqid>:
+  64:	bf041000 	beaeqid	r4, 4096		// 1064
+
+0000000000000068 <beagei>:
+  68:	bda41000 	beagei	r4, 4096		// 1068
+
+000000000000006c <beageid>:
+  6c:	bfa41000 	beageid	r4, 4096		// 106c
+
+0000000000000070 <beagti>:
+  70:	bd841000 	beagti	r4, 4096		// 1070
+
+0000000000000074 <beagtid>:
+  74:	bf841000 	beagtid	r4, 4096		// 1074
+
+0000000000000078 <bealei>:
+  78:	bd641000 	bealei	r4, 4096		// 1078
+
+000000000000007c <bealeid>:
+  7c:	bf641000 	bealeid	r4, 4096		// 107c
+
+0000000000000080 <bealti>:
+  80:	bd441000 	bealti	r4, 4096		// 1080
+
+0000000000000084 <bealtid>:
+  84:	bf441000 	bealtid	r4, 4096		// 1084
+
+0000000000000088 <beanei>:
+  88:	bd241000 	beanei	r4, 4096		// 1088
+
+000000000000008c <beaneid>:
+  8c:	bf241000 	beaneid	r4, 4096		// 108c
+
+0000000000000090 <brea>:
+  90:	98012000 	brea	r4
+
+0000000000000094 <bread>:
+  94:	98112000 	bread	r4
+
+0000000000000098 <breald>:
+  98:	98953800 	breald	r4, r7
+
+000000000000009c <breai>:
+  9c:	b8011000 	breai	4096		// 109c
+
+00000000000000a0 <breaid>:
+  a0:	b8111000 	breaid	4096		// 10a0
+
+00000000000000a4 <brealid>:
+  a4:	b8951000 	brealid	r4, 4096
diff --git a/gas/testsuite/gas/microblaze/brea.exp b/gas/testsuite/gas/microblaze/brea.exp
new file mode 100644
index 00000000000..54d64b92116
--- /dev/null
+++ b/gas/testsuite/gas/microblaze/brea.exp
@@ -0,0 +1,21 @@
+# Copyright (C) 2012-2024 Free Software Foundation, Inc.
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street - Fifth Floor, Boston, MA 02110-1301, USA.
+
+# MicroBlaze test for special register.
+
+if [istarget microblaze*-*-*] {
+    run_dump_test "brea"
+}
diff --git a/gas/testsuite/gas/microblaze/brea.s b/gas/testsuite/gas/microblaze/brea.s
new file mode 100644
index 00000000000..df6d9874f38
--- /dev/null
+++ b/gas/testsuite/gas/microblaze/brea.s
@@ -0,0 +1,129 @@
+    .text
+footext:
+    .text
+	.global beaeq
+beaeq:
+	beaeq r4,r7
+	.global bealeq
+bealeq:
+	bealeq r4,r7
+	.global beaeqd
+beaeqd:
+	beaeqd r4,r7
+	.global bealeqd
+bealeqd:
+	bealeqd r4,r7
+	.global beage
+beage:
+	beage r4,r7
+	.global bealge
+bealge:
+	bealge r4,r7
+	.global beaged
+beaged:
+	beaged r4,r7
+	.global bealged
+bealged:
+	bealged r4,r7
+	.global beagt
+beagt:
+	beagt r4,r7
+	.global bealgt
+bealgt:
+	bealgt r4,r7
+	.global beagtd
+beagtd:
+	beagtd r4,r7
+	.global bealgtd
+bealgtd:
+	bealgtd r4,r7
+	.global beale
+beale:
+	beale r4,r7
+	.global bealle
+bealle:
+	bealle r4,r7
+	.global bealed
+bealed:
+	bealed r4,r7
+	.global bealled
+bealled:
+	bealled r4,r7
+	.global bealt
+bealt:
+	bealt r4,r7
+	.global beallt
+beallt:
+	beallt r4,r7
+	.global bealtd
+bealtd:
+	bealtd r4,r7
+	.global bealltd
+bealltd:
+	bealltd r4,r7
+	.global beane
+beane:
+	beane r4,r7
+	.global bealne
+bealne:
+	bealne r4,r7
+	.global beaned
+beaned:
+	beaned r4,r7
+	.global bealned
+bealned:
+	bealned r4,r7
+	.global beaeqi
+beaeqi:
+	beaeqi r4, 0x1000
+	.global beaeqid
+beaeqid:
+	beaeqid r4, 0x1000
+	.global beagei
+beagei:
+	beagei r4, 0x1000
+	.global beageid
+beageid:
+	beageid r4, 0x1000
+	.global beagti
+beagti:
+	beagti r4, 0x1000
+	.global beagtid
+beagtid:
+	beagtid r4, 0x1000
+	.global bealei
+bealei:
+	bealei r4, 0x1000
+	.global bealeid
+bealeid:
+	bealeid r4, 0x1000
+	.global bealti
+bealti:
+	bealti r4, 0x1000
+	.global bealtid
+bealtid:
+	bealtid r4, 0x1000
+	.global beanei
+beanei:
+	beanei r4, 0x1000
+	.global beaneid
+beaneid:
+	beaneid r4, 0x1000
+	.global brea
+brea:
+	brea r4
+	.global bread
+bread:
+	bread r4
+	.global breald
+breald:
+	breald r4, r7
+	.global breai
+breai:
+	breai 0x1000
+	.global breaid
+breaid:
+	breaid 0x1000
+	.global brealid
+brealid:
+	brealid r4, 0x1000
diff --git a/gas/testsuite/gas/microblaze/reloc_64.d b/gas/testsuite/gas/microblaze/reloc_64.d
new file mode 100644
index 00000000000..3a72ae332e6
--- /dev/null
+++ b/gas/testsuite/gas/microblaze/reloc_64.d
@@ -0,0 +1,23 @@
+#as: -m64
+#objdump: -rd
+
+.*: +file format .*
+
+Disassembly of section .text:
+
+0000000000000000 <i_am_right_here>:
+   0:	6824ffc0 	addlik	r1, -64
+   4:	fde10000 	sli	r15, r1, 0
+   8:	fe610038 	sli	r19, r1, 56
+   c:	12610100 	addlk	r19, r1, r0
+  10:	b2000000 	imml	0
+			10: R_MICROBLAZE_64_PCREL	far_away_function
+  14:	b9f50000 	brealid	r15, 0
+  18:	80000000 	or	r0, r0, r0
+  1c:	80000000 	or	r0, r0, r0
+  20:	ede10000 	lli	r15, r1, 0
+  24:	10330100 	addlk	r1, r19, r0
+  28:	ee610038 	lli	r19, r1, 56
+  2c:	68240040 	addlik	r1, 64
+  30:	b60f0008 	rtsd	r15, 8
+  34:	80000000 	or	r0, r0, r0
diff --git a/gas/testsuite/gas/microblaze/reloc_64.exp b/gas/testsuite/gas/microblaze/reloc_64.exp
new file mode 100644
index 00000000000..1007ea96ac6
--- /dev/null
+++ b/gas/testsuite/gas/microblaze/reloc_64.exp
@@ -0,0 +1,21 @@
+# Copyright (C) 2012-2024 Free Software Foundation, Inc.
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street - Fifth Floor, Boston, MA 02110-1301, USA.
+
+# MicroBlaze test for special register.
+
+if [istarget microblaze*-*-*] {
+    run_dump_test "reloc_64"
+}
diff --git a/gas/testsuite/gas/microblaze/reloc_64.s b/gas/testsuite/gas/microblaze/reloc_64.s
new file mode 100644
index 00000000000..264ef5930fc
--- /dev/null
+++ b/gas/testsuite/gas/microblaze/reloc_64.s
@@ -0,0 +1,26 @@
+	.text
+	.align	3
+	.globl	i_am_right_here
+	.ent	i_am_right_here
+	.type	i_am_right_here, @function
+i_am_right_here:
+	.frame	r19,64,r15		# vars= 0, regs= 1, args= 48
+	.mask	0x00088000
+	addlik	r1,-64
+	sli	r15,r1,0
+	sli	r19,r1,56
+	addlk	r19,r1,r0
+	brealid	r15,far_away_function
+	nop		# Unfilled delay slot
+
+	nop
+	lli	r15,r1,0
+	addlk	r1,r19,r0
+	lli	r19,r1,56
+	addlik	r1,64
+	rtsd	r15,8
+	nop		# Unfilled delay slot
+
+	.end	i_am_right_here
+$Lfe1:
+	.size	i_am_right_here,$Lfe1-i_am_right_here
diff --git a/ld/Makefile.am b/ld/Makefile.am
index c219662d2e3..6a44e1bee76 100644
--- a/ld/Makefile.am
+++ b/ld/Makefile.am
@@ -418,6 +418,8 @@ ALL_64_EMULATION_SOURCES = \
 	eelf32ltsmipn32.c \
 	eelf32ltsmipn32_fbsd.c \
 	eelf32mipswindiss.c \
+        eelf64microblazeel.c \
+        eelf64microblaze.c \
 	eelf64_aix.c \
 	eelf64_ia64.c \
 	eelf64_ia64_fbsd.c \
diff --git a/ld/configure.tgt b/ld/configure.tgt
index 3e158913b89..2c587842a2f 100644
--- a/ld/configure.tgt
+++ b/ld/configure.tgt
@@ -506,7 +506,10 @@ microblaze*-linux*)	targ_emul="elf32mb_linux"
 			targ_extra_emuls="elf32mbel_linux"
 			;;
 microblazeel*)		targ_emul=elf32microblazeel
-			targ_extra_emuls=elf32microblaze
+			targ_extra_emuls="elf32microblaze elf64microblazeel elf64microblaze"
+			;;
+microblazeel64*)	targ_emul=elf64microblazeel
+			targ_extra_emuls=elf64microblaze
 			;;
 microblaze*)		targ_emul=elf32microblaze
 			targ_extra_emuls=elf32microblazeel
diff --git a/ld/emulparams/elf64microblaze.sh b/ld/emulparams/elf64microblaze.sh
new file mode 100644
index 00000000000..7b4c7c411bd
--- /dev/null
+++ b/ld/emulparams/elf64microblaze.sh
@@ -0,0 +1,23 @@
+SCRIPT_NAME=elfmicroblaze
+OUTPUT_FORMAT="elf64-microblazeel"
+#BIG_OUTPUT_FORMAT="elf64-microblaze"
+LITTLE_OUTPUT_FORMAT="elf64-microblazeel"
+#TEXT_START_ADDR=0
+NONPAGED_TEXT_START_ADDR=0x28
+ALIGNMENT=4
+MAXPAGESIZE=4
+ARCH=microblaze
+EMBEDDED=yes
+
+NOP=0x80000000
+
+# Hmmm, there's got to be a better way.  This sets the stack to the
+# top of the simulator memory (2^19 bytes).
+#PAGE_SIZE=0x1000
+#DATA_ADDR=0x10000
+#OTHER_RELOCATING_SECTIONS='.stack 0x7000 : { _stack = .; *(.stack) }'
+#$@{RELOCATING+ PROVIDE (__stack = 0x7000);@}
+#OTHER_RELOCATING_SECTIONS='PROVIDE (_stack = _end + 0x1000);'
+
+TEMPLATE_NAME=elf
+#GENERATE_SHLIB_SCRIPT=yes
diff --git a/ld/emulparams/elf64microblazeel.sh b/ld/emulparams/elf64microblazeel.sh
new file mode 100644
index 00000000000..7b4c7c411bd
--- /dev/null
+++ b/ld/emulparams/elf64microblazeel.sh
@@ -0,0 +1,23 @@
+SCRIPT_NAME=elfmicroblaze
+OUTPUT_FORMAT="elf64-microblazeel"
+#BIG_OUTPUT_FORMAT="elf64-microblaze"
+LITTLE_OUTPUT_FORMAT="elf64-microblazeel"
+#TEXT_START_ADDR=0
+NONPAGED_TEXT_START_ADDR=0x28
+ALIGNMENT=4
+MAXPAGESIZE=4
+ARCH=microblaze
+EMBEDDED=yes
+
+NOP=0x80000000
+
+# Hmmm, there's got to be a better way.  This sets the stack to the
+# top of the simulator memory (2^19 bytes).
+#PAGE_SIZE=0x1000
+#DATA_ADDR=0x10000
+#OTHER_RELOCATING_SECTIONS='.stack 0x7000 : { _stack = .; *(.stack) }'
+#$@{RELOCATING+ PROVIDE (__stack = 0x7000);@}
+#OTHER_RELOCATING_SECTIONS='PROVIDE (_stack = _end + 0x1000);'
+
+TEMPLATE_NAME=elf
+#GENERATE_SHLIB_SCRIPT=yes
diff --git a/ld/testsuite/ld-microblaze/microblaze64.exp b/ld/testsuite/ld-microblaze/microblaze64.exp
new file mode 100644
index 00000000000..72cdfae250b
--- /dev/null
+++ b/ld/testsuite/ld-microblaze/microblaze64.exp
@@ -0,0 +1,26 @@
+# Expect script for ld-microblaze tests
+#   Copyright (C) 2008-2025 Free Software Foundation, Inc.
+#
+# This file is part of the GNU Binutils.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street - Fifth Floor, Boston,
+# MA 02110-1301, USA.
+#
+
+if { !([istarget "*microblaze*-*-*"]) } {
+    return
+}
+
+run_dump_test "s64"
diff --git a/ld/testsuite/ld-microblaze/s64.d b/ld/testsuite/ld-microblaze/s64.d
new file mode 100644
index 00000000000..b50b840c232
--- /dev/null
+++ b/ld/testsuite/ld-microblaze/s64.d
@@ -0,0 +1,66 @@
+#name: MicroBlaze64
+#source: s64.s
+#source: s64defs.s
+#as: -m64
+#ld: -melf64microblazeel -T simple.ld
+#objdump: -d
+
+# Test 64-bit relocations
+.*: +file format elf64-microblazeel
+
+
+Disassembly of section .text:
+
+0000000000000000 <i_am_right_here>:
+   0:	6824ffc8 	addlik	r1, -56
+   4:	fde10000 	sli	r15, r1, 0
+   8:	b2000000 	imml	0
+   c:	b9f57fe4 	brealid	r15, 32740	// 7ff0 <function16>
+  10:	80000000 	or	r0, r0, r0
+  14:	b2001fff 	imml	8191
+  18:	b9f5ffd8 	brealid	r15, -40	// fffffffffffffff0 <function64>
+  1c:	80000000 	or	r0, r0, r0
+  20:	b20fffff 	imml	1048575
+  24:	b9f5ffcc 	brealid	r15, -52	// fffffffffffffff0 <function64>
+  28:	80000000 	or	r0, r0, r0
+  2c:	b2ffffff 	imml	16777215
+  30:	b9f5ffc0 	brealid	r15, -64	// fffffffffffffff0 <function64>
+  34:	80000000 	or	r0, r0, r0
+  38:	ede10000 	lli	r15, r1, 0
+  3c:	68240038 	addlik	r1, 56
+  40:	b60f0008 	rtsd	r15, 8
+  44:	80000000 	or	r0, r0, r0
+
+0000000000000048 <far_fetch>:
+  48:	b2001fff 	imml	8191
+  4c:	e880fff8 	lwi	r4, r0, -8
+  50:	b2000000 	imml	0
+  54:	e8607ff8 	lwi	r3, r0, 32760	// 7ff8 <i16>
+  58:	10632000 	addk	r3, r3, r4
+  5c:	b20fffff 	imml	1048575
+  60:	e880fff8 	lwi	r4, r0, -8
+  64:	10632000 	addk	r3, r3, r4
+  68:	b2ffffff 	imml	16777215
+  6c:	e880fff8 	lwi	r4, r0, -8
+  70:	10632000 	addk	r3, r3, r4
+  74:	b2000000 	imml	0
+  78:	f86000bc 	swi	r3, r0, 188	// bc <g>
+  7c:	b60f0008 	rtsd	r15, 8
+  80:	80000000 	or	r0, r0, r0
+
+Disassembly of section .text.startup:
+
+0000000000000088 <main>:
+  88:	6824ffc8 	addlik	r1, -56
+  8c:	fde10000 	sli	r15, r1, 0
+  90:	b2ffffff 	imml	16777215
+  94:	b9f5ff6c 	brealid	r15, -148	// 0 <i_am_right_here>
+  98:	80000000 	or	r0, r0, r0
+  9c:	b2ffffff 	imml	16777215
+  a0:	b9f5ffa8 	brealid	r15, -88	// 48 <far_fetch>
+  a4:	80000000 	or	r0, r0, r0
+  a8:	ede10000 	lli	r15, r1, 0
+  ac:	10600000 	addk	r3, r0, r0
+  b0:	68240038 	addlik	r1, 56
+  b4:	b60f0008 	rtsd	r15, 8
+  b8:	80000000 	or	r0, r0, r0
diff --git a/ld/testsuite/ld-microblaze/s64.s b/ld/testsuite/ld-microblaze/s64.s
new file mode 100644
index 00000000000..8d48c2b0da7
--- /dev/null
+++ b/ld/testsuite/ld-microblaze/s64.s
@@ -0,0 +1,82 @@
+	.text
+	.align	3
+	.globl	_start
+_start:
+	.globl	i_am_right_here
+	.ent	i_am_right_here
+	.type	i_am_right_here, @function
+i_am_right_here:
+	.frame	r1,56,r15		# vars= 0, regs= 0, args= 48
+	.mask	0x00008000
+	addlik	r1,-56
+	sli	r15,r1,0
+	brealid	r15,function16
+	nop		# Unfilled delay slot
+
+	brealid	r15,function32
+	nop		# Unfilled delay slot
+
+	brealid	r15,function36
+	nop		# Unfilled delay slot
+
+	brealid	r15,function40
+	nop		# Unfilled delay slot
+
+	lli	r15,r1,0
+	addlik	r1,56
+	rtsd	r15,8
+	nop		# Unfilled delay slot
+
+	.end	i_am_right_here
+$Lfe1:
+	.size	i_am_right_here,$Lfe1-i_am_right_here
+	.align	3
+	.globl	far_fetch
+	.ent	far_fetch
+	.type	far_fetch, @function
+far_fetch:
+	.frame	r1,0,r15		# vars= 0, regs= 0, args= 0
+	.mask	0x00000000
+	lwi	r4,r0,i32
+	lwi	r3,r0,i16
+	addk	r3,r3,r4
+	lwi	r4,r0,i36
+	addk	r3,r3,r4
+	lwi	r4,r0,i40
+	addk	r3,r3,r4
+	swi	r3,r0,g
+	rtsd	r15,8
+	nop		# Unfilled delay slot
+
+	.end	far_fetch
+$Lfe2:
+	.size	far_fetch,$Lfe2-far_fetch
+	.section	.text.startup,"ax",@progbits
+	.align	3
+	.globl	main
+	.ent	main
+	.type	main, @function
+main:
+	.frame	r1,56,r15		# vars= 0, regs= 0, args= 48
+	.mask	0x00008000
+	addlik	r1,-56
+	sli	r15,r1,0
+	brealid	r15,i_am_right_here
+	nop		# Unfilled delay slot
+
+	brealid	r15,far_fetch
+	nop		# Unfilled delay slot
+
+	lli	r15,r1,0
+	addk	r3,r0,r0
+	addlik	r1,56
+	rtsd	r15,8
+	nop		# Unfilled delay slot
+
+	.end	main
+$Lfe3:
+	.size	main,$Lfe3-main
+	.globl	g
+	.bss
+	.lcomm	g,4,4
+	.type	g, @object
diff --git a/ld/testsuite/ld-microblaze/s64defs.s b/ld/testsuite/ld-microblaze/s64defs.s
new file mode 100644
index 00000000000..391d9b92d77
--- /dev/null
+++ b/ld/testsuite/ld-microblaze/s64defs.s
@@ -0,0 +1,27 @@
+.global function16
+.global function32
+.global function36
+.global function40
+.global function44
+.global function64
+.global i16
+.global i32
+.global i36
+.global i40
+.global i44
+.global i64
+
+
+.set function16, 0x7ff0
+.set function32, 0x1ffffff0
+.set function36, 0xffffffff0
+.set function40, 0xfffffffff0
+.set function44, 0xffffffffff0
+.set function64, 0xfffffffffffffff0
+
+.set i16, 0x7ff8
+.set i32, 0x1ffffff8
+.set i36, 0xffffffff8
+.set i40, 0xfffffffff8
+.set i44, 0xffffffffff8
+.set i64, 0xfffffffffffffff0
diff --git a/ld/testsuite/ld-microblaze/simple.ld b/ld/testsuite/ld-microblaze/simple.ld
new file mode 100644
index 00000000000..3f9f7c5cde1
--- /dev/null
+++ b/ld/testsuite/ld-microblaze/simple.ld
@@ -0,0 +1,9 @@
+ENTRY(_start)
+SECTIONS
+{
+  . = 0;
+  .text : {
+    *(.text)
+  }
+   _etext  =  .;
+}
-- 
2.47.3

